{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Creating enzyme-substrate database from GO Annotation database for Uniprot IDs\n",
    "\n",
    "### 1. Extracting all GO Terms with information about chemical reactions\n",
    "### 2. Getting IDs for the substrates\n",
    "### 3. Searching the GO Annotations database for UniProt IDs for entries with GO Terms stating the catalyitic activity of an enzyme\n",
    "### 4. Splitting the dataset in training and testing set\n",
    "### 5. Calculating enzyme and substrate representations\n",
    "### 6. Adding negative data points\n",
    "### 7. Adding task-specific enzyme representations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C:\\Users\\alexk\\projects\\SubFinder\\notebooks_and_code\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "from os.path import join\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import time\n",
    "from rdkit import Chem\n",
    "from rdkit import DataStructs\n",
    "from rdkit.Chem import AllChem\n",
    "from Bio import SeqIO\n",
    "import warnings\n",
    "import torch\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "sys.path.append('.\\\\additional_code')\n",
    "from data_preprocessing import *\n",
    "\n",
    "CURRENT_DIR = os.getcwd()\n",
    "print(CURRENT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Extract all GO Terms with information about chemical reactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Create pandas DataFrame with alle GO Terms (with infomration about the catalytic activity of enzymes) including its names, definitions, and reaction IDs:\n",
    "go.obo file with information about GO Terms was downloaded from http://geneontology.org/docs/download-ontology/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Name</th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>\"Catalysis of the reaction: all-trans-hexapren...</td>\n",
       "      <td>trans-hexaprenyltranstransferase activity</td>\n",
       "      <td>20836</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>\"Catalysis of the reaction: lactose + H2O = D-...</td>\n",
       "      <td>lactase activity</td>\n",
       "      <td>10076</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>\"Catalysis of the reaction: adenine + H2O = hy...</td>\n",
       "      <td>adenine deaminase activity</td>\n",
       "      <td>23688</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000048</td>\n",
       "      <td>\"Catalysis of the reaction: peptidyl-tRNA(1) +...</td>\n",
       "      <td>peptidyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000104</td>\n",
       "      <td>\"Catalysis of the reaction: succinate + accept...</td>\n",
       "      <td>succinate dehydrogenase activity</td>\n",
       "      <td>16357</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GO ID                                         Definition  \\\n",
       "0  GO:0000010  \"Catalysis of the reaction: all-trans-hexapren...   \n",
       "1  GO:0000016  \"Catalysis of the reaction: lactose + H2O = D-...   \n",
       "2  GO:0000034  \"Catalysis of the reaction: adenine + H2O = hy...   \n",
       "3  GO:0000048  \"Catalysis of the reaction: peptidyl-tRNA(1) +...   \n",
       "4  GO:0000104  \"Catalysis of the reaction: succinate + accept...   \n",
       "\n",
       "                                        Name RHEA ID KEGG ID  \n",
       "0  trans-hexaprenyltranstransferase activity   20836     NaN  \n",
       "1                           lactase activity   10076     NaN  \n",
       "2                 adenine deaminase activity   23688     NaN  \n",
       "3               peptidyltransferase activity     NaN     NaN  \n",
       "4           succinate dehydrogenase activity   16357     NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(columns = [\"GO ID\", \"Definition\", \"Name\", \"RHEA ID\"])\n",
    "\n",
    "#read go obo file\n",
    "path_go_obo = join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", 'go.obo')\n",
    "file = open(path_go_obo, 'r')\n",
    "Lines = file.read()\n",
    "\n",
    "#process data from go.obo-file and save it in a pandas DataFrame:\n",
    "start = Lines.index('[Term]\\n')\n",
    "while start != -1:\n",
    "    start = Lines.index('[Term]\\n')\n",
    "    Lines = Lines[start+1:]\n",
    "    try:\n",
    "        GO_Term = Lines[: Lines.index('[Term]\\n')]\n",
    "    except ValueError:\n",
    "        start = -1\n",
    "    definition, name, ID = get_info_from_GO_Term(GO_Term = GO_Term)\n",
    "    if \"Catalysis of the reaction\" in definition:\n",
    "        RHEA_ID= get_RHEA_reaction_IDs(GO_ID = ID, GO_Term = GO_Term)\n",
    "        df = df.append({\"GO ID\" : ID , \"Definition\" : definition, \"Name\" : name,\n",
    "                        \"RHEA ID\" : RHEA_ID}, ignore_index = True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"df_GO_catalytic.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Name</th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>\"Catalysis of the reaction: all-trans-hexapren...</td>\n",
       "      <td>trans-hexaprenyltranstransferase activity</td>\n",
       "      <td>20836</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>\"Catalysis of the reaction: lactose + H2O = D-...</td>\n",
       "      <td>lactase activity</td>\n",
       "      <td>10076</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>\"Catalysis of the reaction: adenine + H2O = hy...</td>\n",
       "      <td>adenine deaminase activity</td>\n",
       "      <td>23688</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000048</td>\n",
       "      <td>\"Catalysis of the reaction: peptidyl-tRNA(1) +...</td>\n",
       "      <td>peptidyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000104</td>\n",
       "      <td>\"Catalysis of the reaction: succinate + accept...</td>\n",
       "      <td>succinate dehydrogenase activity</td>\n",
       "      <td>16357</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6582</th>\n",
       "      <td>GO:1990833</td>\n",
       "      <td>\"Catalysis of the reaction: ATP + H2O = ADP + ...</td>\n",
       "      <td>clathrin-uncoating ATPase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6583</th>\n",
       "      <td>GO:1990883</td>\n",
       "      <td>\"Catalysis of the reaction: acetyl-CoA + cytid...</td>\n",
       "      <td>rRNA cytidine N-acetyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6584</th>\n",
       "      <td>GO:1990887</td>\n",
       "      <td>\"Catalysis of the reaction: 2-polyprenyl-3-met...</td>\n",
       "      <td>2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6585</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>\"Catalysis of the reaction: 2-polyprenyl-6-hyd...</td>\n",
       "      <td>2-polyprenyl-6-hydroxyphenol O-methyltransfera...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6586</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>\"Catalysis of the reaction: cytosylglucuronic ...</td>\n",
       "      <td>cytosylglucuronate decarboxylase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6587 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GO ID                                         Definition  \\\n",
       "0     GO:0000010  \"Catalysis of the reaction: all-trans-hexapren...   \n",
       "1     GO:0000016  \"Catalysis of the reaction: lactose + H2O = D-...   \n",
       "2     GO:0000034  \"Catalysis of the reaction: adenine + H2O = hy...   \n",
       "3     GO:0000048  \"Catalysis of the reaction: peptidyl-tRNA(1) +...   \n",
       "4     GO:0000104  \"Catalysis of the reaction: succinate + accept...   \n",
       "...          ...                                                ...   \n",
       "6582  GO:1990833  \"Catalysis of the reaction: ATP + H2O = ADP + ...   \n",
       "6583  GO:1990883  \"Catalysis of the reaction: acetyl-CoA + cytid...   \n",
       "6584  GO:1990887  \"Catalysis of the reaction: 2-polyprenyl-3-met...   \n",
       "6585  GO:1990888  \"Catalysis of the reaction: 2-polyprenyl-6-hyd...   \n",
       "6586  GO:1990965  \"Catalysis of the reaction: cytosylglucuronic ...   \n",
       "\n",
       "                                                   Name RHEA ID KEGG ID  \n",
       "0             trans-hexaprenyltranstransferase activity   20836     NaN  \n",
       "1                                      lactase activity   10076     NaN  \n",
       "2                            adenine deaminase activity   23688     NaN  \n",
       "3                          peptidyltransferase activity     NaN     NaN  \n",
       "4                      succinate dehydrogenase activity   16357     NaN  \n",
       "...                                                 ...     ...     ...  \n",
       "6582                 clathrin-uncoating ATPase activity     NaN     NaN  \n",
       "6583         rRNA cytidine N-acetyltransferase activity     NaN     NaN  \n",
       "6584  2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4-...     NaN     NaN  \n",
       "6585  2-polyprenyl-6-hydroxyphenol O-methyltransfera...     NaN     NaN  \n",
       "6586          cytosylglucuronate decarboxylase activity     NaN     NaN  \n",
       "\n",
       "[6587 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_catalytic = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"df_GO_catalytic.pkl\"))\n",
    "df_catalytic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Getting IDs for the substrates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Trying to get substrate name via text mining of the GO Term definitions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not find substrate in the following definition: which results in unsaturation at C-7 in the B ring of sterols\n",
      "Could not find substrate in the following definition: GlcNAc-1,6-anhMurNAc-L-Ala-gamma-D-Glu-DAP-D-Ala + H2O glcNAc-1,6-anhMurNAc + L-Ala-gamma-D-Glu-DAP-D-Ala\n",
      "Could not find substrate in the following definition: involving the transfer of a phosphatidate (otherwise known as diacylglycerol 3-phosphosphate) group\n",
      "Could not find substrate in the following definition: ATP + undecaprenol + all-trans-undecaprenyl phosphate + ADP + H+\n",
      "Could not find substrate in the following definition: 7-hydroxymethyl chlorophyll a + 2 reduced ferredoxin + 2 H+ chlorophyll a + 2 oxidized ferredoxin + H2O\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Name</th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>substrates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>\"Catalysis of the reaction: all-trans-hexapren...</td>\n",
       "      <td>trans-hexaprenyltranstransferase activity</td>\n",
       "      <td>20836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[all-trans-hexaprenyl diphosphate, isopentenyl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>\"Catalysis of the reaction: lactose + H2O = D-...</td>\n",
       "      <td>lactase activity</td>\n",
       "      <td>10076</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[lactose, H2O ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>\"Catalysis of the reaction: adenine + H2O = hy...</td>\n",
       "      <td>adenine deaminase activity</td>\n",
       "      <td>23688</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[adenine, H2O ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000048</td>\n",
       "      <td>\"Catalysis of the reaction: peptidyl-tRNA(1) +...</td>\n",
       "      <td>peptidyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[peptidyl-tRNA(1), aminoacyl-tRNA(2) ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000104</td>\n",
       "      <td>\"Catalysis of the reaction: succinate + accept...</td>\n",
       "      <td>succinate dehydrogenase activity</td>\n",
       "      <td>16357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[succinate, acceptor ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6582</th>\n",
       "      <td>GO:1990833</td>\n",
       "      <td>\"Catalysis of the reaction: ATP + H2O = ADP + ...</td>\n",
       "      <td>clathrin-uncoating ATPase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ATP, H2O ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6583</th>\n",
       "      <td>GO:1990883</td>\n",
       "      <td>\"Catalysis of the reaction: acetyl-CoA + cytid...</td>\n",
       "      <td>rRNA cytidine N-acetyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[acetyl-CoA, cytidine ]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6584</th>\n",
       "      <td>GO:1990887</td>\n",
       "      <td>\"Catalysis of the reaction: 2-polyprenyl-3-met...</td>\n",
       "      <td>2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6585</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>\"Catalysis of the reaction: 2-polyprenyl-6-hyd...</td>\n",
       "      <td>2-polyprenyl-6-hydroxyphenol O-methyltransfera...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[2-polyprenyl-6-hydroxyphenol, S-adenosyl-L-me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6586</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>\"Catalysis of the reaction: cytosylglucuronic ...</td>\n",
       "      <td>cytosylglucuronate decarboxylase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cytosylglucuronic acid, H(+) ]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6587 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GO ID                                         Definition  \\\n",
       "0     GO:0000010  \"Catalysis of the reaction: all-trans-hexapren...   \n",
       "1     GO:0000016  \"Catalysis of the reaction: lactose + H2O = D-...   \n",
       "2     GO:0000034  \"Catalysis of the reaction: adenine + H2O = hy...   \n",
       "3     GO:0000048  \"Catalysis of the reaction: peptidyl-tRNA(1) +...   \n",
       "4     GO:0000104  \"Catalysis of the reaction: succinate + accept...   \n",
       "...          ...                                                ...   \n",
       "6582  GO:1990833  \"Catalysis of the reaction: ATP + H2O = ADP + ...   \n",
       "6583  GO:1990883  \"Catalysis of the reaction: acetyl-CoA + cytid...   \n",
       "6584  GO:1990887  \"Catalysis of the reaction: 2-polyprenyl-3-met...   \n",
       "6585  GO:1990888  \"Catalysis of the reaction: 2-polyprenyl-6-hyd...   \n",
       "6586  GO:1990965  \"Catalysis of the reaction: cytosylglucuronic ...   \n",
       "\n",
       "                                                   Name RHEA ID KEGG ID  \\\n",
       "0             trans-hexaprenyltranstransferase activity   20836     NaN   \n",
       "1                                      lactase activity   10076     NaN   \n",
       "2                            adenine deaminase activity   23688     NaN   \n",
       "3                          peptidyltransferase activity     NaN     NaN   \n",
       "4                      succinate dehydrogenase activity   16357     NaN   \n",
       "...                                                 ...     ...     ...   \n",
       "6582                 clathrin-uncoating ATPase activity     NaN     NaN   \n",
       "6583         rRNA cytidine N-acetyltransferase activity     NaN     NaN   \n",
       "6584  2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4-...     NaN     NaN   \n",
       "6585  2-polyprenyl-6-hydroxyphenol O-methyltransfera...     NaN     NaN   \n",
       "6586          cytosylglucuronate decarboxylase activity     NaN     NaN   \n",
       "\n",
       "                                             substrates  \n",
       "0     [all-trans-hexaprenyl diphosphate, isopentenyl...  \n",
       "1                                       [lactose, H2O ]  \n",
       "2                                       [adenine, H2O ]  \n",
       "3                [peptidyl-tRNA(1), aminoacyl-tRNA(2) ]  \n",
       "4                                [succinate, acceptor ]  \n",
       "...                                                 ...  \n",
       "6582                                        [ATP, H2O ]  \n",
       "6583                            [acetyl-CoA, cytidine ]  \n",
       "6584  [2-polyprenyl-3-methyl-5-hydroxy-6-methoxy-1,4...  \n",
       "6585  [2-polyprenyl-6-hydroxyphenol, S-adenosyl-L-me...  \n",
       "6586                    [cytosylglucuronic acid, H(+) ]  \n",
       "\n",
       "[6587 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_catalytic[\"substrates\"] = \"\"\n",
    "\n",
    "for ind in df_catalytic.index:\n",
    "    df_catalytic[\"substrates\"][ind] = find_substrates(definition = df_catalytic[\"Definition\"][ind])\n",
    "    \n",
    "df_catalytic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (b) Getting substrate CHEBI IDs for the reactions with RHEA IDs:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(i) Mapping the metabolites of the reaction with RHEA IDs to CHEBI metabolite IDs\n",
    "Information about RHEA reactions were downloaded from https://ftp.expasy.org/databases/rhea/txt/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>CHEBI IDs</th>\n",
       "      <th>CHEBI_ID_list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:15377, CHEBI:16459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:15377, CHEBI:16459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10002.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:28938, CHEBI:31011]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10003.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:15377, CHEBI:16459]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10004.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:17484]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54195</th>\n",
       "      <td>66623.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:71550, CHEBI:15379, CHEBI:57618]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54196</th>\n",
       "      <td>66624.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:16175, CHEBI:15379, CHEBI:57618]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54197</th>\n",
       "      <td>66625.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:16175, CHEBI:15379, CHEBI:57618]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54198</th>\n",
       "      <td>66626.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:71541, CHEBI:15378, CHEBI:15377, CHEBI:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54199</th>\n",
       "      <td>66627.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:16175, CHEBI:15379, CHEBI:57618]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54200 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       RHEA ID CHEBI IDs                                      CHEBI_ID_list\n",
       "0      10000.0       NaN                         [CHEBI:15377, CHEBI:16459]\n",
       "1      10001.0       NaN                         [CHEBI:15377, CHEBI:16459]\n",
       "2      10002.0       NaN                         [CHEBI:28938, CHEBI:31011]\n",
       "3      10003.0       NaN                         [CHEBI:15377, CHEBI:16459]\n",
       "4      10004.0       NaN                                      [CHEBI:17484]\n",
       "...        ...       ...                                                ...\n",
       "54195  66623.0       NaN            [CHEBI:71550, CHEBI:15379, CHEBI:57618]\n",
       "54196  66624.0       NaN            [CHEBI:16175, CHEBI:15379, CHEBI:57618]\n",
       "54197  66625.0       NaN            [CHEBI:16175, CHEBI:15379, CHEBI:57618]\n",
       "54198  66626.0       NaN  [CHEBI:71541, CHEBI:15378, CHEBI:15377, CHEBI:...\n",
       "54199  66627.0       NaN            [CHEBI:16175, CHEBI:15379, CHEBI:57618]\n",
       "\n",
       "[54200 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_RHEA = pd.DataFrame(columns = [\"RHEA ID\", \"CHEBI IDs\", \"CHEBI_ID_list\"])\n",
    "\n",
    "file1 = open(join(CURRENT_DIR, \"..\" ,\"data\", \"reaction_data\", \"rhea-reactions.txt\"), 'r')\n",
    "Lines = file1.readlines()\n",
    "\n",
    "while True:\n",
    "    try:\n",
    "        end = Lines.index('///\\n')\n",
    "        entry = Lines[:end]\n",
    "        RHEA_ID, CHEBI_IDs = extract_RHEA_ID_and_CHEBI_IDs(entry)\n",
    "        CHEBI_ID_list = get_substrate_IDs(IDs = CHEBI_IDs)\n",
    "        Lines = Lines[end+1:]\n",
    "        df_RHEA = df_RHEA.append({\"RHEA ID\" : RHEA_ID, \"CHEBI_ID_list\" : CHEBI_ID_list}, ignore_index = True)\n",
    "    except ValueError:\n",
    "        break\n",
    "        \n",
    "df_RHEA[\"RHEA ID\"] = [float(ID.split(\":\")[-1]) for ID in df_RHEA[\"RHEA ID\"]]\n",
    "df_RHEA.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"reaction_data\", \"RHEA_reaction_df.pkl\"))\n",
    "df_RHEA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(ii) Mapping CHEBI IDs to df_catalytic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Name</th>\n",
       "      <th>RHEA ID</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>substrates</th>\n",
       "      <th>CHEBI IDs</th>\n",
       "      <th>CHEBI_ID_list</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>\"Catalysis of the reaction: all-trans-hexapren...</td>\n",
       "      <td>trans-hexaprenyltranstransferase activity</td>\n",
       "      <td>20836.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[all-trans-hexaprenyl diphosphate, isopentenyl...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:58179, CHEBI:128769]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>\"Catalysis of the reaction: lactose + H2O = D-...</td>\n",
       "      <td>lactase activity</td>\n",
       "      <td>10076.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[lactose, H2O ]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:15377, CHEBI:17716]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>\"Catalysis of the reaction: adenine + H2O = hy...</td>\n",
       "      <td>adenine deaminase activity</td>\n",
       "      <td>23688.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[adenine, H2O ]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:16708, CHEBI:15378, CHEBI:15377]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000048</td>\n",
       "      <td>\"Catalysis of the reaction: peptidyl-tRNA(1) +...</td>\n",
       "      <td>peptidyltransferase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[peptidyl-tRNA(1), aminoacyl-tRNA(2) ]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000104</td>\n",
       "      <td>\"Catalysis of the reaction: succinate + accept...</td>\n",
       "      <td>succinate dehydrogenase activity</td>\n",
       "      <td>16357.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[succinate, acceptor ]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CHEBI:13193, CHEBI:30031]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GO ID                                         Definition  \\\n",
       "0  GO:0000010  \"Catalysis of the reaction: all-trans-hexapren...   \n",
       "1  GO:0000016  \"Catalysis of the reaction: lactose + H2O = D-...   \n",
       "2  GO:0000034  \"Catalysis of the reaction: adenine + H2O = hy...   \n",
       "3  GO:0000048  \"Catalysis of the reaction: peptidyl-tRNA(1) +...   \n",
       "4  GO:0000104  \"Catalysis of the reaction: succinate + accept...   \n",
       "\n",
       "                                        Name  RHEA ID KEGG ID  \\\n",
       "0  trans-hexaprenyltranstransferase activity  20836.0     NaN   \n",
       "1                           lactase activity  10076.0     NaN   \n",
       "2                 adenine deaminase activity  23688.0     NaN   \n",
       "3               peptidyltransferase activity      NaN     NaN   \n",
       "4           succinate dehydrogenase activity  16357.0     NaN   \n",
       "\n",
       "                                          substrates CHEBI IDs  \\\n",
       "0  [all-trans-hexaprenyl diphosphate, isopentenyl...       NaN   \n",
       "1                                    [lactose, H2O ]       NaN   \n",
       "2                                    [adenine, H2O ]       NaN   \n",
       "3             [peptidyl-tRNA(1), aminoacyl-tRNA(2) ]       NaN   \n",
       "4                             [succinate, acceptor ]       NaN   \n",
       "\n",
       "                             CHEBI_ID_list  \n",
       "0              [CHEBI:58179, CHEBI:128769]  \n",
       "1               [CHEBI:15377, CHEBI:17716]  \n",
       "2  [CHEBI:16708, CHEBI:15378, CHEBI:15377]  \n",
       "3                                      NaN  \n",
       "4               [CHEBI:13193, CHEBI:30031]  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_catalytic[\"RHEA ID\"] = [float(ID) for ID in df_catalytic[\"RHEA ID\"]]\n",
    "df_catalytic = df_catalytic.merge(df_RHEA, how = \"left\", on = [\"RHEA ID\"])\n",
    "df_catalytic.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(iii) Creating DataFrame with all combinations of GO Terms and single substrates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>molecule</th>\n",
       "      <th>molecule ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:58179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:128769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:17716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:16708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14649</th>\n",
       "      <td>GO:1990887</td>\n",
       "      <td>S-adenosyl-L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14650</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>2-polyprenyl-6-hydroxyphenol</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14651</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>S-adenosyl-L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14652</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>cytosylglucuronic acid</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14653</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>H(+)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14654 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GO ID                      molecule   molecule ID\n",
       "0      GO:0000010                           NaN   CHEBI:58179\n",
       "1      GO:0000010                           NaN  CHEBI:128769\n",
       "2      GO:0000016                           NaN   CHEBI:15377\n",
       "3      GO:0000016                           NaN   CHEBI:17716\n",
       "4      GO:0000034                           NaN   CHEBI:16708\n",
       "...           ...                           ...           ...\n",
       "14649  GO:1990887      S-adenosyl-L-methionine            NaN\n",
       "14650  GO:1990888  2-polyprenyl-6-hydroxyphenol           NaN\n",
       "14651  GO:1990888      S-adenosyl-L-methionine            NaN\n",
       "14652  GO:1990965        cytosylglucuronic acid           NaN\n",
       "14653  GO:1990965                         H(+)            NaN\n",
       "\n",
       "[14654 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_substrates = pd.DataFrame(columns = [\"GO ID\", \"molecule\", \"molecule ID\"])\n",
    "df_catalytic[\"CHEBI_ID_list\"].loc[pd.isnull(df_catalytic[\"CHEBI_ID_list\"])] = \"\"\n",
    "\n",
    "for ind in df_catalytic.index:\n",
    "    GO_ID = df_catalytic[\"GO ID\"][ind]\n",
    "    \n",
    "    if len(df_catalytic[\"CHEBI_ID_list\"][ind]) > 0:\n",
    "        IDs = df_catalytic[\"CHEBI_ID_list\"][ind]\n",
    "        for ID in IDs:\n",
    "            df_substrates = df_substrates.append({\"GO ID\" : GO_ID, \"molecule\": np.nan, \"molecule ID\" : ID},\n",
    "                                                 ignore_index = True)\n",
    "    else:\n",
    "        metabolites = df_catalytic[\"substrates\"][ind]\n",
    "        for met in metabolites:\n",
    "            df_substrates = df_substrates.append({\"GO ID\" : GO_ID, \"molecule\": met, \"molecule ID\" : np.nan},\n",
    "                                                 ignore_index = True)\n",
    "            \n",
    "df_substrates.reset_index(inplace = True, drop = True)\n",
    "df_substrates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  (c) Trying to map all substrates without a CHEBI ID to an identifier:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>xanthan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>indan-1-ol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>oxidosqualene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>galactogen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(histone)-arginine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2402</th>\n",
       "      <td>GDP-mannose = mannan(n+1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2403</th>\n",
       "      <td>a 3-oxo-dodecanoyl-[acp]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2404</th>\n",
       "      <td>coenzyme A or its derivatives</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405</th>\n",
       "      <td>DNA with alkylated base</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2406</th>\n",
       "      <td>2,4-dihydroxy-hept-trans-2-ene-1,7-dioate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2407 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     metabolites\n",
       "0                                       xanthan \n",
       "1                                     indan-1-ol\n",
       "2                                 oxidosqualene \n",
       "3                                     galactogen\n",
       "4                            (histone)-arginine \n",
       "...                                          ...\n",
       "2402                   GDP-mannose = mannan(n+1)\n",
       "2403                   a 3-oxo-dodecanoyl-[acp] \n",
       "2404               coenzyme A or its derivatives\n",
       "2405                     DNA with alkylated base\n",
       "2406  2,4-dihydroxy-hept-trans-2-ene-1,7-dioate \n",
       "\n",
       "[2407 rows x 1 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metabolites = []\n",
    "for ind in df_substrates.index:\n",
    "    if pd.isnull(df_substrates[\"molecule ID\"][ind]):\n",
    "        metabolites = metabolites + [df_substrates[\"molecule\"][ind]]\n",
    "    \n",
    "df_unmapped = pd.DataFrame(data = {\"metabolites\" : list(set(metabolites))})\n",
    "df_unmapped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c)(i) Mapping metabolite names to KEGG compound synonym database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For 1589 out of 2407 data points we could not map the substrate name to a KEGG ID.\n"
     ]
    }
   ],
   "source": [
    "datasets_dir = \"C:\\\\Users\\\\alexk\\\\projects\\\\Prediction_of_KM_V4\\\\datasets\\\\\"\n",
    "\n",
    "drugs_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"KEGG_drugs_df.pkl\"))\n",
    "compounds_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"KEGG_substrate_df.pkl\"))\n",
    "KEGG_substrate_df = compounds_df.append(drugs_df).reset_index(drop = True)\n",
    "\n",
    "##If we have multiple IDs for the same substrate name, we keep the first ID:\n",
    "droplist = []\n",
    "for ind in KEGG_substrate_df.index:\n",
    "    if not ind in droplist:\n",
    "        substrate = KEGG_substrate_df[\"substrate\"][ind]\n",
    "        help_df = KEGG_substrate_df.loc[KEGG_substrate_df[\"substrate\"] == substrate]\n",
    "        if len(help_df) > 1 :\n",
    "            droplist = droplist + list(help_df.index)[1:]\n",
    "\n",
    "KEGG_substrate_df.drop(droplist, inplace = True)\n",
    "\n",
    "KEGG_substrate_df[\"substrate\"] = [name.lower() for name in KEGG_substrate_df[\"substrate\"]]\n",
    "df_unmapped[\"substrate\"] = [name.lower() for name in df_unmapped[\"metabolites\"]]\n",
    "\n",
    "df_unmapped = df_unmapped.merge(KEGG_substrate_df, on = \"substrate\", how = \"left\")\n",
    "print(\"For %s out of %s data points we could not map the substrate name to a KEGG ID.\" %\n",
    "      (sum(pd.isnull(df_unmapped[\"KEGG ID\"])), len(df_unmapped)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c)(ii) Mapping all substrate names without a KEGG CID to PubChem CIDs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "unmapped_substrates = list(set(list(df_unmapped[\"substrate\"][pd.isnull(list(df_unmapped[\"KEGG ID\"]))])))\n",
    "matches = substrate_names_to_Pubchem_CIDs(unmapped_substrates)\n",
    "matches.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"Pubchem_substrate_matches_GO_Terms.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mapping all PubChem CIDs to KEGG CIDs: Therefore, we create a txt-file (\"Pubchem_CIDs_GO_Terms.txt\") that contains all PubChem CIDs that we want to map to KEGG CIDs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert CIDs from strings to integers:\n",
    "for ind in matches.index:\n",
    "    cid = matches.loc[ind][\"CID\"]\n",
    "    if not pd.isnull(cid):\n",
    "        matches[\"CID\"][ind] = int(cid)\n",
    "        \n",
    "        \n",
    "#create txt file with all CIDs in matches:\n",
    "CIDs = list(matches.loc[~ pd.isnull(matches[\"CID\"])][\"CID\"])\n",
    "f = open(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"Pubchem_CIDs_GO_Terms.txt\"),\"w\") \n",
    "for cid in CIDs:\n",
    "    f.write(str(cid) + \"\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The txt-file \"Pubchem_CIDs_GO_Terms.txt\" can be used as the input for the webservice https://www.metaboanalyst.ca/MetaboAnalyst/upload/ConvertView.xhtml to map the Pubchem CIDs to KEGG CIDs and CHEBI IDs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Metabolite</th>\n",
       "      <th>CID</th>\n",
       "      <th>ChEBI</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>METLIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>coash</td>\n",
       "      <td>87642.0</td>\n",
       "      <td>15346.0</td>\n",
       "      <td>C00010</td>\n",
       "      <td>6235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dehydroabietadienal</td>\n",
       "      <td>11694869.0</td>\n",
       "      <td>52487.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tricaffeoyl spermidine</td>\n",
       "      <td>15241070.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chlorophyllide b(1-)</td>\n",
       "      <td>86289090.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>beta-chaconine</td>\n",
       "      <td>119393.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>2,4-dinitrocyclohexanone</td>\n",
       "      <td>9543365.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>2-(methylthio)benzothiazole</td>\n",
       "      <td>11989.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>indole acetic acid</td>\n",
       "      <td>802.0</td>\n",
       "      <td>16411.0</td>\n",
       "      <td>C00954</td>\n",
       "      <td>5211.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>(r)-3-hydroxyicosanoyl-coa(4-)</td>\n",
       "      <td>72193813.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>monocyclic sesterterpenediol</td>\n",
       "      <td>90658711.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>452 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Metabolite         CID    ChEBI    KEGG  METLIN\n",
       "0                             coash     87642.0  15346.0  C00010  6235.0\n",
       "1               dehydroabietadienal  11694869.0  52487.0     NaN     NaN\n",
       "2            tricaffeoyl spermidine  15241070.0      NaN     NaN     NaN\n",
       "3              chlorophyllide b(1-)  86289090.0      NaN     NaN     NaN\n",
       "4                    beta-chaconine    119393.0      NaN     NaN     NaN\n",
       "..                              ...         ...      ...     ...     ...\n",
       "447        2,4-dinitrocyclohexanone   9543365.0      NaN     NaN     NaN\n",
       "448     2-(methylthio)benzothiazole     11989.0      NaN     NaN     NaN\n",
       "449              indole acetic acid       802.0  16411.0  C00954  5211.0\n",
       "450  (r)-3-hydroxyicosanoyl-coa(4-)  72193813.0      NaN     NaN     NaN\n",
       "451    monocyclic sesterterpenediol  90658711.0      NaN     NaN     NaN\n",
       "\n",
       "[452 rows x 5 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matches= pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"Pubchem_substrate_matches_GO_Terms.pkl\"))\n",
    "matches[\"CID\"] = [int(cid) if not pd.isnull(cid) else np.nan for cid in matches[\"CID\"]]\n",
    "matches.head()\n",
    "matches = matches.loc[~pd.isnull(matches[\"CID\"])]\n",
    "\n",
    "#load the resulting file and store it in a DataFrame:\n",
    "Pubchem_CID_to_KEGG_df = pd.read_csv(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"name_map_metaboanalyst.csv\"), sep = \",\")\n",
    "#rename columns:\n",
    "Pubchem_CID_to_KEGG_df.rename(columns = {\"PubChem\" : \"CID\"}, inplace = True)\n",
    "Pubchem_CID_to_KEGG_df.drop(columns = [\"Match\",\"Query\", \"HMDB\", \"SMILES\", \"Comment\"], inplace = True)\n",
    "\n",
    "#merge this DataFrame with the DataFrame called matches:\n",
    "macthes_with_KEGG_IDs = pd.merge(matches, Pubchem_CID_to_KEGG_df, how='left', on=['CID'])\n",
    "macthes_with_KEGG_IDs.rename(columns = {\"Match\" : \"substrate\"}, inplace = True)\n",
    "macthes_with_KEGG_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>substrate</th>\n",
       "      <th>KEGG ID</th>\n",
       "      <th>CHEBI ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>xanthan</td>\n",
       "      <td>xanthan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>indan-1-ol</td>\n",
       "      <td>indan-1-ol</td>\n",
       "      <td>C01710</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>oxidosqualene</td>\n",
       "      <td>oxidosqualene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>galactogen</td>\n",
       "      <td>galactogen</td>\n",
       "      <td>C01698</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(histone)-arginine</td>\n",
       "      <td>(histone)-arginine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2402</th>\n",
       "      <td>GDP-mannose = mannan(n+1)</td>\n",
       "      <td>gdp-mannose = mannan(n+1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2403</th>\n",
       "      <td>a 3-oxo-dodecanoyl-[acp]</td>\n",
       "      <td>a 3-oxo-dodecanoyl-[acp]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2404</th>\n",
       "      <td>coenzyme A or its derivatives</td>\n",
       "      <td>coenzyme a or its derivatives</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2405</th>\n",
       "      <td>DNA with alkylated base</td>\n",
       "      <td>dna with alkylated base</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2406</th>\n",
       "      <td>2,4-dihydroxy-hept-trans-2-ene-1,7-dioate</td>\n",
       "      <td>2,4-dihydroxy-hept-trans-2-ene-1,7-dioate</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2407 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     metabolites  \\\n",
       "0                                       xanthan    \n",
       "1                                     indan-1-ol   \n",
       "2                                 oxidosqualene    \n",
       "3                                     galactogen   \n",
       "4                            (histone)-arginine    \n",
       "...                                          ...   \n",
       "2402                   GDP-mannose = mannan(n+1)   \n",
       "2403                   a 3-oxo-dodecanoyl-[acp]    \n",
       "2404               coenzyme A or its derivatives   \n",
       "2405                     DNA with alkylated base   \n",
       "2406  2,4-dihydroxy-hept-trans-2-ene-1,7-dioate    \n",
       "\n",
       "                                       substrate KEGG ID  CHEBI ID  \n",
       "0                                       xanthan      NaN       NaN  \n",
       "1                                     indan-1-ol  C01710       NaN  \n",
       "2                                 oxidosqualene      NaN       NaN  \n",
       "3                                     galactogen  C01698       NaN  \n",
       "4                            (histone)-arginine      NaN       NaN  \n",
       "...                                          ...     ...       ...  \n",
       "2402                   gdp-mannose = mannan(n+1)     NaN       NaN  \n",
       "2403                   a 3-oxo-dodecanoyl-[acp]      NaN       NaN  \n",
       "2404               coenzyme a or its derivatives     NaN       NaN  \n",
       "2405                     dna with alkylated base     NaN       NaN  \n",
       "2406  2,4-dihydroxy-hept-trans-2-ene-1,7-dioate      NaN       NaN  \n",
       "\n",
       "[2407 rows x 4 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_unmapped[\"CHEBI ID\"] = np.nan\n",
    "\n",
    "for ind in df_unmapped.index:\n",
    "    if pd.isnull(df_unmapped[\"KEGG ID\"][ind]):\n",
    "        substrate = df_unmapped[\"substrate\"][ind]\n",
    "        try:\n",
    "            KEGG_ID = list(macthes_with_KEGG_IDs.loc[macthes_with_KEGG_IDs[\"substrate\"]== substrate][\"KEGG ID\"])[0]\n",
    "            df_unmapped[\"KEGG ID\"][ind] = KEGG_ID\n",
    "        except:\n",
    "            None\n",
    "        try:\n",
    "            CHEBI_ID = list(macthes_with_KEGG_IDs.loc[macthes_with_KEGG_IDs[\"substrate\"]== substrate][\"ChEBI\"])[0]\n",
    "            df_unmapped[\"CHEBI ID\"][ind] = CHEBI_ID\n",
    "        except:\n",
    "            None\n",
    "df_unmapped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c)(iii) Mapping substrate IDs to df_substrate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>molecule</th>\n",
       "      <th>molecule ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:58179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:128769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:17716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CHEBI:16708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14649</th>\n",
       "      <td>GO:1990887</td>\n",
       "      <td>S-adenosyl-L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14650</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>2-polyprenyl-6-hydroxyphenol</td>\n",
       "      <td>C17551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14651</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>S-adenosyl-L-methionine</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14652</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>cytosylglucuronic acid</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14653</th>\n",
       "      <td>GO:1990965</td>\n",
       "      <td>H(+)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14654 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GO ID                      molecule   molecule ID\n",
       "0      GO:0000010                           NaN   CHEBI:58179\n",
       "1      GO:0000010                           NaN  CHEBI:128769\n",
       "2      GO:0000016                           NaN   CHEBI:15377\n",
       "3      GO:0000016                           NaN   CHEBI:17716\n",
       "4      GO:0000034                           NaN   CHEBI:16708\n",
       "...           ...                           ...           ...\n",
       "14649  GO:1990887      S-adenosyl-L-methionine            NaN\n",
       "14650  GO:1990888  2-polyprenyl-6-hydroxyphenol        C17551\n",
       "14651  GO:1990888      S-adenosyl-L-methionine            NaN\n",
       "14652  GO:1990965        cytosylglucuronic acid           NaN\n",
       "14653  GO:1990965                         H(+)            NaN\n",
       "\n",
       "[14654 rows x 3 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for ind in df_substrates.index:\n",
    "    if pd.isnull(df_substrates[\"molecule ID\"][ind]):\n",
    "        met = df_substrates[\"molecule\"][ind]\n",
    "\n",
    "        ID = list(df_unmapped[\"CHEBI ID\"].loc[df_unmapped[\"metabolites\"] == met])[0]\n",
    "        if pd.isnull(ID):\n",
    "            ID = list(df_unmapped[\"KEGG ID\"].loc[df_unmapped[\"metabolites\"] == met])[0]\n",
    "\n",
    "        df_substrates[\"molecule ID\"][ind] = ID\n",
    "df_substrates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>molecule ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>CHEBI:58179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000010</td>\n",
       "      <td>CHEBI:128769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0000016</td>\n",
       "      <td>CHEBI:17716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0000034</td>\n",
       "      <td>CHEBI:16708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11304</th>\n",
       "      <td>GO:1990738</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11305</th>\n",
       "      <td>GO:1990738</td>\n",
       "      <td>CHEBI:58380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11306</th>\n",
       "      <td>GO:1990833</td>\n",
       "      <td>C00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11307</th>\n",
       "      <td>GO:1990883</td>\n",
       "      <td>C00024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11308</th>\n",
       "      <td>GO:1990888</td>\n",
       "      <td>C17551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11309 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            GO ID   molecule ID\n",
       "0      GO:0000010   CHEBI:58179\n",
       "1      GO:0000010  CHEBI:128769\n",
       "2      GO:0000016   CHEBI:15377\n",
       "3      GO:0000016   CHEBI:17716\n",
       "4      GO:0000034   CHEBI:16708\n",
       "...           ...           ...\n",
       "11304  GO:1990738   CHEBI:15377\n",
       "11305  GO:1990738   CHEBI:58380\n",
       "11306  GO:1990833        C00002\n",
       "11307  GO:1990883        C00024\n",
       "11308  GO:1990888        C17551\n",
       "\n",
       "[11309 rows x 2 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_substrates.drop(columns = [\"molecule\"], inplace= True)\n",
    "df_substrates = df_substrates.loc[~pd.isnull(df_substrates[\"molecule ID\"])]\n",
    "df_substrates.reset_index(inplace = True, drop = True)\n",
    "df_substrates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_substrates.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"df_GO_catalytic.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Searching the GO Annotations database for entries with GO Terms stating the catalyitic activity of an enzyme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_substrates = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"df_GO_catalytic.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a) Searching the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The GO Annotation database contains evidence codes (ECOs). These ECOs contain the information which kind of evidence an annotation has, e.g. \"computational\", \"experimential\", \"phylogenetic\",...\n",
    "We are only interested in data points with experimental and phylogenetic evidence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"df_GO_catalytic.pkl\"))\n",
    "catalytic_go_terms = list(set(df[\"GO ID\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_GO_UID = pd.read_csv(join(CURRENT_DIR, \"alex_data\", \"df_GO_UID.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5763"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(catalytic_go_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ECO_to_GAF = pd.read_csv(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", 'ECO_to_GAF.tsv'), sep = \"\\t\")\n",
    "exp_evidence = [\"EXP\",\"IDA\",\"IPI\",\"IMP\",\"IGI\",\"IEP\", \"HTP\",\"HDA\",\"HMP\",\"HGI\",\"HEP\"]\n",
    "phylo_evidence = [\"IBA\",\"IBD\",\"IKR\",\"IRD\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The GOA database contains over 922 million entries. We created a function \"search_GOA_database\" to parallize the search. It takes the argument run which should be an integer betwenn 0 and 922 and searches one million entries accoriding to the value run has.\n",
    "The GOA database for uniprot IDs was downloaded from here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for run in range(923):\n",
    "#    search_GOA_database(run)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)Mapping GO_terms with phylogenetic and experimental evidence to substrate IDs:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merging the created DataFrames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_GO_UID = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"experimental_and_phylogenetic\",\n",
    "                                \"experimental_phylogenetic_df_GO_UID_part_\" + str(0) +\".pkl\"))\n",
    "\n",
    "for i in range(1,923):\n",
    "    try:\n",
    "        df_new = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"experimental_and_phylogenetic\",\n",
    "                                     \"experimental_phylogenetic_df_GO_UID_part_\" + str(i) +\".pkl\"))\n",
    "        df_GO_UID = pd.concat([df_GO_UID, df_new], ignore_index= True)\n",
    "    except FileNotFoundError:\n",
    "        print(\"Error\", i)\n",
    "df_GO_UID.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"experimental_phylogenetic_df_GO_UID.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_GO_UID = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"GOA_data\", \"experimental_phylogenetic_df_GO_UID.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mapping Uniprot IDs and metabolite IDs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "1001\n",
      "2001\n",
      "3001\n",
      "4001\n",
      "5001\n",
      "6001\n",
      "7001\n",
      "8001\n",
      "9001\n",
      "10001\n",
      "11001\n",
      "12001\n",
      "13001\n",
      "14001\n",
      "15001\n",
      "16001\n",
      "17001\n",
      "18001\n",
      "19001\n",
      "20001\n",
      "21001\n",
      "22001\n",
      "23001\n",
      "24001\n",
      "25001\n",
      "26001\n",
      "27001\n",
      "28001\n",
      "29001\n",
      "30001\n",
      "31001\n",
      "32001\n",
      "33001\n",
      "34001\n",
      "35001\n",
      "36001\n",
      "37001\n",
      "38001\n",
      "39001\n",
      "40001\n",
      "41001\n",
      "42001\n",
      "43001\n",
      "44001\n",
      "45001\n",
      "46001\n",
      "47001\n",
      "48001\n",
      "49001\n",
      "50001\n",
      "51001\n",
      "52001\n",
      "53001\n",
      "54001\n",
      "55001\n",
      "56001\n",
      "57001\n",
      "58001\n",
      "59001\n",
      "60001\n",
      "61001\n",
      "62001\n",
      "63001\n",
      "64001\n",
      "65001\n",
      "66001\n",
      "67001\n",
      "68001\n",
      "69001\n",
      "70001\n",
      "71001\n",
      "72001\n",
      "73001\n",
      "74001\n",
      "75001\n",
      "76001\n",
      "77001\n",
      "78001\n",
      "79001\n",
      "80001\n",
      "81001\n",
      "82001\n",
      "83001\n",
      "84001\n",
      "85001\n",
      "86001\n",
      "87001\n",
      "88001\n",
      "89001\n",
      "90001\n",
      "91001\n",
      "92001\n",
      "93001\n",
      "94001\n",
      "95001\n",
      "96001\n",
      "97001\n",
      "98001\n",
      "99001\n",
      "100001\n",
      "101001\n",
      "102001\n",
      "103001\n",
      "104001\n",
      "105001\n",
      "106001\n",
      "107001\n",
      "108001\n",
      "109001\n",
      "110001\n",
      "111001\n",
      "112001\n",
      "113001\n",
      "114001\n",
      "115001\n",
      "116001\n",
      "117001\n",
      "118001\n",
      "119001\n",
      "120001\n",
      "121001\n",
      "122001\n",
      "123001\n",
      "124001\n",
      "125001\n",
      "126001\n",
      "127001\n",
      "128001\n",
      "129001\n",
      "130001\n",
      "131001\n",
      "132001\n",
      "133001\n",
      "134001\n",
      "135001\n",
      "136001\n",
      "137001\n",
      "138001\n",
      "139001\n",
      "140001\n",
      "141001\n",
      "142001\n",
      "143001\n",
      "144001\n",
      "145001\n",
      "146001\n",
      "147001\n",
      "148001\n",
      "149001\n",
      "150001\n",
      "151001\n",
      "152001\n",
      "153001\n",
      "154001\n",
      "155001\n",
      "156001\n",
      "157001\n",
      "158001\n",
      "159001\n",
      "160001\n",
      "161001\n",
      "162001\n",
      "163001\n",
      "164001\n",
      "165001\n",
      "166001\n",
      "167001\n",
      "168001\n",
      "169001\n",
      "170001\n",
      "171001\n",
      "172001\n",
      "173001\n",
      "174001\n",
      "175001\n",
      "176001\n",
      "177001\n",
      "178001\n",
      "179001\n",
      "180001\n",
      "181001\n",
      "182001\n",
      "183001\n",
      "184001\n",
      "185001\n",
      "186001\n",
      "187001\n",
      "188001\n",
      "189001\n",
      "190001\n",
      "191001\n",
      "192001\n",
      "193001\n",
      "194001\n",
      "195001\n",
      "196001\n",
      "197001\n",
      "198001\n",
      "199001\n",
      "200001\n",
      "201001\n",
      "202001\n",
      "203001\n",
      "204001\n",
      "205001\n",
      "206001\n",
      "207001\n",
      "208001\n",
      "209001\n",
      "210001\n",
      "211001\n",
      "212001\n",
      "213001\n",
      "214001\n",
      "215001\n",
      "216001\n",
      "217001\n",
      "218001\n",
      "219001\n",
      "220001\n",
      "221001\n",
      "222001\n",
      "223001\n",
      "224001\n",
      "225001\n",
      "226001\n",
      "227001\n",
      "228001\n",
      "229001\n",
      "230001\n",
      "231001\n",
      "232001\n",
      "233001\n",
      "234001\n",
      "235001\n",
      "236001\n",
      "237001\n",
      "238001\n",
      "239001\n",
      "240001\n",
      "241001\n",
      "242001\n",
      "243001\n",
      "244001\n",
      "245001\n",
      "246001\n",
      "247001\n",
      "248001\n",
      "249001\n",
      "250001\n",
      "251001\n",
      "252001\n",
      "253001\n",
      "254001\n",
      "255001\n",
      "256001\n",
      "257001\n",
      "258001\n",
      "259001\n",
      "260001\n",
      "261001\n",
      "262001\n",
      "263001\n",
      "264001\n",
      "265001\n",
      "266001\n",
      "267001\n",
      "268001\n",
      "269001\n",
      "270001\n",
      "271001\n",
      "272001\n",
      "273001\n",
      "274001\n",
      "275001\n",
      "276001\n",
      "277001\n",
      "278001\n",
      "279001\n",
      "236843 1772\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>molecule ID</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>CHEBI:57540</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A022PMU5</td>\n",
       "      <td>C00030</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A022PN36</td>\n",
       "      <td>CHEBI:57318</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A022PN36</td>\n",
       "      <td>CHEBI:57540</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495070</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>CHEBI:15378</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495071</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>CHEBI:15379</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495072</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>CHEBI:57394</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495073</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>C00154</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495074</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>C00030</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>462406 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID  molecule ID evidence\n",
       "0       A0A009IHW8  CHEBI:15377      exp\n",
       "1       A0A009IHW8  CHEBI:57540      exp\n",
       "2       A0A022PMU5       C00030    phylo\n",
       "3       A0A022PN36  CHEBI:57318    phylo\n",
       "4       A0A022PN36  CHEBI:57540    phylo\n",
       "...            ...          ...      ...\n",
       "495070      Z4YNJ9  CHEBI:15378    phylo\n",
       "495071      Z4YNJ9  CHEBI:15379    phylo\n",
       "495072      Z4YNJ9  CHEBI:57394    phylo\n",
       "495073      Z4YNJ9       C00154    phylo\n",
       "495074      Z4YNJ9       C00030    phylo\n",
       "\n",
       "[462406 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_UID_MID = pd.DataFrame(columns =[\"Uniprot ID\", \"molecule ID\", \"evidence\"])\n",
    "\n",
    "for ind in df_GO_UID.index:\n",
    "    if ind >= -1:\n",
    "        GO_ID = df_GO_UID[\"GO Term\"][ind]\n",
    "        UID = df_GO_UID[\"Uniprot ID\"][ind]\n",
    "        evidence = df_GO_UID[\"evidence\"][ind]\n",
    "        met_IDs = list(df_substrates[\"molecule ID\"].loc[df_substrates[\"GO ID\"] == GO_ID])\n",
    "        for met_ID in met_IDs:\n",
    "            df_UID_MID = df_UID_MID.append({\"Uniprot ID\" : UID, \"molecule ID\" : met_ID,\n",
    "                                           \"evidence\": evidence}, ignore_index = True)\n",
    "    if ind % 1000 ==1:\n",
    "        print(ind)\n",
    "        \n",
    "df_UID_MID.drop_duplicates(inplace = True)\n",
    "df_UID_MID.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID.pkl\"))\n",
    "\n",
    "Uniprot_IDs = list(set(df_UID_MID[\"Uniprot ID\"]))\n",
    "print(len(Uniprot_IDs), len(list(set(df_UID_MID[\"molecule ID\"]))))\n",
    "\n",
    "df_UID_MID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>molecule ID</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>CHEBI:15377</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>CHEBI:57540</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10167</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CHEBI:16731</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10168</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CHEBI:57287</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10169</th>\n",
       "      <td>A0A059TC02</td>\n",
       "      <td>CHEBI:58349</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495053</th>\n",
       "      <td>X5KCU3</td>\n",
       "      <td>C00535</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495054</th>\n",
       "      <td>X5KCU9</td>\n",
       "      <td>C00535</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495055</th>\n",
       "      <td>X5KJC0</td>\n",
       "      <td>C00535</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495056</th>\n",
       "      <td>X5L1L5</td>\n",
       "      <td>C00535</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495057</th>\n",
       "      <td>X5L565</td>\n",
       "      <td>C00535</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28112 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID  molecule ID evidence\n",
       "0       A0A009IHW8  CHEBI:15377      exp\n",
       "1       A0A009IHW8  CHEBI:57540      exp\n",
       "10167   A0A059TC02  CHEBI:16731      exp\n",
       "10168   A0A059TC02  CHEBI:57287      exp\n",
       "10169   A0A059TC02  CHEBI:58349      exp\n",
       "...            ...          ...      ...\n",
       "495053      X5KCU3       C00535      exp\n",
       "495054      X5KCU9       C00535      exp\n",
       "495055      X5KJC0       C00535      exp\n",
       "495056      X5L1L5       C00535      exp\n",
       "495057      X5L565       C00535      exp\n",
       "\n",
       "[28112 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_UID_MID.loc[df_UID_MID[\"evidence\"] == \"exp\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are removing small molecules like H2O and H+ from the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>molecule ID</th>\n",
       "      <th>evidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>CHEBI:57540</td>\n",
       "      <td>exp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A022PMU5</td>\n",
       "      <td>C00030</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A022PN36</td>\n",
       "      <td>CHEBI:57318</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A022PN36</td>\n",
       "      <td>CHEBI:57540</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A0A022PN36</td>\n",
       "      <td>CHEBI:16469</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495066</th>\n",
       "      <td>Z4YHZ5</td>\n",
       "      <td>CHEBI:16810</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495067</th>\n",
       "      <td>Z4YHZ5</td>\n",
       "      <td>CHEBI:50342</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495072</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>CHEBI:57394</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495073</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>C00154</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495074</th>\n",
       "      <td>Z4YNJ9</td>\n",
       "      <td>C00030</td>\n",
       "      <td>phylo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>369461 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID  molecule ID evidence\n",
       "1       A0A009IHW8  CHEBI:57540      exp\n",
       "2       A0A022PMU5       C00030    phylo\n",
       "3       A0A022PN36  CHEBI:57318    phylo\n",
       "4       A0A022PN36  CHEBI:57540    phylo\n",
       "6       A0A022PN36  CHEBI:16469    phylo\n",
       "...            ...          ...      ...\n",
       "495066      Z4YHZ5  CHEBI:16810    phylo\n",
       "495067      Z4YHZ5  CHEBI:50342    phylo\n",
       "495072      Z4YNJ9  CHEBI:57394    phylo\n",
       "495073      Z4YNJ9       C00154    phylo\n",
       "495074      Z4YNJ9       C00030    phylo\n",
       "\n",
       "[369461 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''#magnesium, Phosphoprotein,  Collagen, C02, O2, Calmodulin, water, RNA, calcium, iron, hydrogen,\n",
    "protein polypeptide chain,DNA, ammonium, lipopolysaccharide, fatty acid anion, double stranded DNA,\n",
    "iron, hydrogen donor, hydrogen acceptor, lipid''';\n",
    "\n",
    "remove_mets = [\"CHEBI:18420\", \"C00562\", \"C00211\", \"CHEBI:16526\", \"CHEBI:15379\", \"C00391\", \"CHEBI:15377\",\n",
    "               \"C00046\", \"CHEBI:29108\", \"CHEBI:29034\", \"CHEBI:15378\", \"CHEBI:16541\", \"CHEBI:16991\",\n",
    "              \"CHEBI:28938\", \"CHEBI:16412\", \"CHEBI:28868\", \"CHEBI:4705\", \"CHEBI:29033\", \"CHEBI:17499\",\n",
    "              \"CHEBI:13193\", \"C01356\"]\n",
    "\n",
    "df_UID_MID = df_UID_MID.loc[~df_UID_MID[\"molecule ID\"].isin(remove_mets)]\n",
    "df_UID_MID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (c) Getting an overview of the number of unique enzymes and substrates in the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have 346923 entries with phylogenetic evidence and 22538 entries with experimental evidence\n",
      "\n",
      " experimental dataset:\n",
      "Number of different enzymes: 13508, Number of different substrates: 1754\n",
      "\n",
      " phylogenetic dataset:\n",
      "Number of different enzymes: 229254, Number of different substrates: 888\n"
     ]
    }
   ],
   "source": [
    "df_exp = df_UID_MID.loc[df_UID_MID[\"evidence\"] == \"exp\"]\n",
    "df_phylo = df_UID_MID.loc[df_UID_MID[\"evidence\"] == \"phylo\"]\n",
    "\n",
    "print(\"We have %s entries with phylogenetic evidence and %s entries with experimental evidence\" % (len(df_phylo), len(df_exp)))\n",
    "\n",
    "print(\"\\n experimental dataset:\")\n",
    "print(\"Number of different enzymes: %s, Number of different substrates: %s\"\n",
    "      % (len(set(df_exp[\"Uniprot ID\"])), len(set(df_exp[\"molecule ID\"]))) )\n",
    "\n",
    "print(\"\\n phylogenetic dataset:\")\n",
    "print(\"Number of different enzymes: %s, Number of different substrates: %s\"\n",
    "      % (len(set(df_phylo[\"Uniprot ID\"])), len(set(df_phylo[\"molecule ID\"]))) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Splitting the dataset in training and testing set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To split the dataset in training and test set, we take the enzyme's sequence identity into account. Therefore, we first cluster the enzymes by sequence identity:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Downloading sequences for all UniProt IDs:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating file with all UniProt IDs: We split it into two files, since UniProt only takes files up to 2MB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "Uniprot_IDs = list(set(df_UID_MID[\"Uniprot ID\"]))\n",
    "\n",
    "Uniprot_IDs_V1 = Uniprot_IDs[:100000]\n",
    "Uniprot_IDs_V2 = Uniprot_IDs[100000:]\n",
    "\n",
    "f = open(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"UNIPROT_IDs_V1.txt\"),\"w\")\n",
    "for ID in list(set(Uniprot_IDs_V1)):\n",
    "    f.write(str(ID) + \"\\n\")\n",
    "f.close()\n",
    "\n",
    "f = open(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"UNIPROT_IDs_V2.txt\"),\"w\")\n",
    "for ID in list(set(Uniprot_IDs_V2)):\n",
    "    f.write(str(ID) + \"\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downloading UniProt IDs via the UniProt mapping service (https://www.uniprot.org/uploadlists/):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-7-dfb95e2dcdb6>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      6\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mfasta\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mfasta_sequences\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m         \u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msequence\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfasta\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfasta\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m         \u001b[0mUNIPROT_df\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mUNIPROT_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;34m\"Uniprot ID\"\u001b[0m \u001b[1;33m:\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"|\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"Sequence\"\u001b[0m \u001b[1;33m:\u001b[0m \u001b[0msequence\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mignore_index\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mappend\u001b[1;34m(self, other, ignore_index, verify_integrity, sort)\u001b[0m\n\u001b[0;32m   7745\u001b[0m             \u001b[0mignore_index\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mignore_index\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   7746\u001b[0m             \u001b[0mverify_integrity\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mverify_integrity\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 7747\u001b[1;33m             \u001b[0msort\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0msort\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   7748\u001b[0m         )\n\u001b[0;32m   7749\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\reshape\\concat.py\u001b[0m in \u001b[0;36mconcat\u001b[1;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[0;32m    285\u001b[0m     )\n\u001b[0;32m    286\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 287\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    288\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    289\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\reshape\\concat.py\u001b[0m in \u001b[0;36mget_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    501\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    502\u001b[0m             new_data = concatenate_block_managers(\n\u001b[1;32m--> 503\u001b[1;33m                 \u001b[0mmgrs_indexers\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnew_axes\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mconcat_axis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbm_axis\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    504\u001b[0m             )\n\u001b[0;32m    505\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\internals\\concat.py\u001b[0m in \u001b[0;36mconcatenate_block_managers\u001b[1;34m(mgrs_indexers, axes, concat_axis, copy)\u001b[0m\n\u001b[0;32m     62\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mview\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m             \u001b[0mb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmake_block_same_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mplacement\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 64\u001b[1;33m         \u001b[1;32melif\u001b[0m \u001b[0m_is_uniform_join_units\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mjoin_units\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     65\u001b[0m             \u001b[0mblk\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mjoin_units\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mblock\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     66\u001b[0m             \u001b[0mvals\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mju\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mblock\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mju\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mjoin_units\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\internals\\concat.py\u001b[0m in \u001b[0;36m_is_uniform_join_units\u001b[1;34m(join_units)\u001b[0m\n\u001b[0;32m    478\u001b[0m         \u001b[1;31m# no blocks that would get missing values (can lead to type upcasts)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    479\u001b[0m         \u001b[1;31m# unless we're an extension dtype.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 480\u001b[1;33m         \u001b[0mall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mnot\u001b[0m \u001b[0mju\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mis_na\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mju\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mblock\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mis_extension\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mju\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mjoin_units\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    481\u001b[0m         \u001b[1;32mand\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    482\u001b[0m         \u001b[1;31m# no blocks with indexers (as then the dimensions do not fit)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\internals\\concat.py\u001b[0m in \u001b[0;36m<genexpr>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m    478\u001b[0m         \u001b[1;31m# no blocks that would get missing values (can lead to type upcasts)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    479\u001b[0m         \u001b[1;31m# unless we're an extension dtype.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 480\u001b[1;33m         \u001b[0mall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mnot\u001b[0m \u001b[0mju\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mis_na\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mju\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mblock\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mis_extension\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mju\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mjoin_units\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    481\u001b[0m         \u001b[1;32mand\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    482\u001b[0m         \u001b[1;31m# no blocks with indexers (as then the dimensions do not fit)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\properties.pyx\u001b[0m in \u001b[0;36mpandas._libs.properties.CachedProperty.__get__\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\internals\\concat.py\u001b[0m in \u001b[0;36mis_na\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    226\u001b[0m         \u001b[0mchunk_len\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtotal_len\u001b[0m \u001b[1;33m//\u001b[0m \u001b[1;36m40\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m1000\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    227\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtotal_len\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mchunk_len\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 228\u001b[1;33m             \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues_flat\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m \u001b[1;33m:\u001b[0m \u001b[0mi\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mchunk_len\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    229\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    230\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\dtypes\\missing.py\u001b[0m in \u001b[0;36misna\u001b[1;34m(obj)\u001b[0m\n\u001b[0;32m    122\u001b[0m     \u001b[0mName\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mbool\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    123\u001b[0m     \"\"\"\n\u001b[1;32m--> 124\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0m_isna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    125\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\dtypes\\missing.py\u001b[0m in \u001b[0;36m_isna\u001b[1;34m(obj, inf_as_na)\u001b[0m\n\u001b[0;32m    155\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    156\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mABCSeries\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mABCIndexClass\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mABCExtensionArray\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 157\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0m_isna_ndarraylike\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minf_as_na\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0minf_as_na\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    158\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mABCDataFrame\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    159\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mobj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\dtypes\\missing.py\u001b[0m in \u001b[0;36m_isna_ndarraylike\u001b[1;34m(obj, inf_as_na)\u001b[0m\n\u001b[0;32m    216\u001b[0m             \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    217\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mis_string_dtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 218\u001b[1;33m         \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_isna_string_dtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minf_as_na\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0minf_as_na\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    219\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mneeds_i8_conversion\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    220\u001b[0m         \u001b[1;31m# this is the NaT pattern\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\envs\\Predicting_Km\\lib\\site-packages\\pandas\\core\\dtypes\\missing.py\u001b[0m in \u001b[0;36m_isna_string_dtype\u001b[1;34m(values, dtype, inf_as_na)\u001b[0m\n\u001b[0;32m    248\u001b[0m             \u001b[0mvec\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlibmissing\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misnaobj\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mravel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    249\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 250\u001b[1;33m         \u001b[0mresult\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m...\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvec\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreshape\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    251\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    252\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "UNIPROT_df = pd.DataFrame(columns = [\"Uniprot ID\", \"Sequence\"])\n",
    "\n",
    "for version in [\"1\", \"2\"]:\n",
    "    fasta_sequences = SeqIO.parse(open(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\",\n",
    "                                            \"UNIPROT_IDs_V\" +version + \".fasta\")),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "        UNIPROT_df = UNIPROT_df.append({\"Uniprot ID\" : name.split(\"|\")[1], \"Sequence\" : sequence}, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"UNIPROT_df.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (b) Clustering enzymes by sequence identity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"UNIPROT_df.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating fasta file with all sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "ofile = open(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", 'clusters', \"all_sequences.fasta\"), \"w\")\n",
    "for ind in UNIPROT_df.index:\n",
    "    seq = UNIPROT_df[\"Sequence\"][ind]\n",
    "    if not pd.isnull(seq):\n",
    "        seq_end = seq.find(\"#\")\n",
    "        seq = seq[:seq_end]\n",
    "        ofile.write(\">\" + str(ind) + \"\\n\" + seq  + \"\\n\")\n",
    "ofile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(i) Clustering:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first cluster in such a way that two different clusters do not contain two enzymes with a sequence identity higher than 80%:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting input for cd-hit algorithm:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster the fasta files\n",
    "cluster_folder = join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", 'clusters')\n",
    "start_folder = cluster_folder\n",
    "cluster_all_levels(start_folder, \n",
    "                   cluster_folder, \n",
    "                   filename='all_sequences')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>230802.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>61688.071425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>37173.511476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>29925.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>59270.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>92366.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>133172.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cluster\n",
       "count  230802.000000\n",
       "mean    61688.071425\n",
       "std     37173.511476\n",
       "min         0.000000\n",
       "25%     29925.250000\n",
       "50%     59270.000000\n",
       "75%     92366.000000\n",
       "max    133172.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>member</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>59618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>17477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>193350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>202039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>31895</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cluster  member\n",
       "0        0   59618\n",
       "1        1   17477\n",
       "2        2  193350\n",
       "3        3  202039\n",
       "4        3   31895"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>member</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>230797</th>\n",
       "      <td>133168</td>\n",
       "      <td>102904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230798</th>\n",
       "      <td>133169</td>\n",
       "      <td>168585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230799</th>\n",
       "      <td>133170</td>\n",
       "      <td>39204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230800</th>\n",
       "      <td>133171</td>\n",
       "      <td>42158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230801</th>\n",
       "      <td>133172</td>\n",
       "      <td>99620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        cluster  member\n",
       "230797   133168  102904\n",
       "230798   133169  168585\n",
       "230799   133170   39204\n",
       "230800   133171   42158\n",
       "230801   133172   99620"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cluster_folder = join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", 'clusters')\n",
    "    \n",
    "# cluster the fasta files\n",
    "start_folder = cluster_folder\n",
    "cluster_all_levels_80(start_folder, \n",
    "                   cluster_folder, \n",
    "                   filename='all_sequences')\n",
    "\n",
    "# collect cluster members\n",
    "df_80 = find_cluster_members_80(folder=cluster_folder, \n",
    "                          filename='all_sequences')\n",
    "\n",
    "display(df_80.describe())\n",
    "display(df_80.head())\n",
    "display(df_80.tail())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We cluster in such a way that two different clusters do not contain two enzymes with a sequence identity higher than 60%:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>230802.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>32736.837146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>21124.000661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>14827.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>29851.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>48600.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>79033.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cluster\n",
       "count  230802.000000\n",
       "mean    32736.837146\n",
       "std     21124.000661\n",
       "min         0.000000\n",
       "25%     14827.000000\n",
       "50%     29851.000000\n",
       "75%     48600.750000\n",
       "max     79033.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cluster_all_levels_60(start_folder, \n",
    "                   cluster_folder, \n",
    "                   filename='all_sequences')\n",
    "\n",
    "# collect cluster members\n",
    "df_60 = find_cluster_members_60(folder=cluster_folder, \n",
    "                       filename='all_sequences')\n",
    "display(df_60.describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first cluster in such a way that two different clusters do not contain two enzymes with a sequence identity higher than 40%:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>230802.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>13197.257376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>9271.144542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>5389.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>11371.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>19518.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>36601.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cluster\n",
       "count  230802.000000\n",
       "mean    13197.257376\n",
       "std      9271.144542\n",
       "min         0.000000\n",
       "25%      5389.000000\n",
       "50%     11371.000000\n",
       "75%     19518.750000\n",
       "max     36601.000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# cluster the fasta files\n",
    "cluster_all_levels(start_folder, \n",
    "                   cluster_folder, \n",
    "                   filename='all_sequences')\n",
    "\n",
    "# collect cluster members\n",
    "df_40 = find_cluster_members(folder=cluster_folder, \n",
    "                          filename='all_sequences')\n",
    "\n",
    "display(df_40.describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.DataFrame({\"member\" : UNIPROT_df.index, \"Uniprot ID\" : UNIPROT_df[\"Uniprot ID\"]})\n",
    "df_UID_MID = df_UID_MID.merge(df2, how = \"left\", on = \"Uniprot ID\")\n",
    "\n",
    "df_40[\"evidence\"] = \"phylo\"\n",
    "for ind in df_40.index:\n",
    "    member = int(df_40[\"member\"][ind])\n",
    "    evidences = list(df_UID_MID[\"evidence\"].loc[df_UID_MID[\"member\"] == member])\n",
    "    if \"exp\" in evidences:\n",
    "        df_40[\"evidence\"][ind] = \"exp\"\n",
    "        \n",
    "        \n",
    "df_60[\"evidence\"] = \"phylo\"\n",
    "for ind in df_60.index:\n",
    "    member = int(df_60[\"member\"][ind])\n",
    "    evidences = list(df_UID_MID[\"evidence\"].loc[df_UID_MID[\"member\"] == member])\n",
    "    if \"exp\" in evidences:\n",
    "        df_60[\"evidence\"][ind] = \"exp\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)(ii) Splitting the dataset in train, validation and test set with a sequence identity cutoff of 80%. Later, we divide the test set in three subparts with identity cutoffs of <40%, 40-60% and 60-80%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df[\"cluster\"] = np.nan\n",
    "for ind in df_80.index:\n",
    "    member = int(df_80[\"member\"][ind])\n",
    "    cluster = df_80[\"cluster\"][ind]\n",
    "    UNIPROT_df[\"cluster\"][member] = cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We only use the additional validation set for the training of the ESM-1b model. It is so big that we can't do a 5-fold CV. Therefore, we use an independent validation set to validate the model instead."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "133185\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(185312, 45514)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = list(set(UNIPROT_df[\"cluster\"]))\n",
    "random.seed(1)\n",
    "random.shuffle(clusters)\n",
    "print(len(clusters))\n",
    "\n",
    "n = int(len(clusters)*0.8)\n",
    "train_clusters = clusters[:n]\n",
    "test_clusters = clusters[n:]\n",
    "\n",
    "training_UIDs = UNIPROT_df[\"Uniprot ID\"].loc[UNIPROT_df[\"cluster\"].isin(train_clusters)]\n",
    "test_UIDs = UNIPROT_df[\"Uniprot ID\"].loc[UNIPROT_df[\"cluster\"].isin(test_clusters)]\n",
    "\n",
    "df_80[\"split\"] = np.nan\n",
    "df_80[\"split\"].loc[df_80[\"cluster\"].isin(train_clusters)] = \"train\"\n",
    "df_80[\"split\"].loc[df_80[\"cluster\"].isin(test_clusters)] = \"test\"\n",
    "\n",
    "train_members = list(df_80[\"member\"].loc[df_80[\"split\"] == \"train\"])\n",
    "test_members = list(df_80[\"member\"].loc[df_80[\"split\"] == \"test\"])\n",
    "\n",
    "df_60[\"split\"] = np.nan\n",
    "df_40[\"split\"] = np.nan\n",
    "df_60[\"split\"].loc[df_60[\"member\"].isin(train_members)] = \"train\"\n",
    "df_60[\"split\"].loc[df_60[\"member\"].isin(test_members)] = \"test\"\n",
    "df_40[\"split\"].loc[df_40[\"member\"].isin(train_members)] = \"train\"\n",
    "df_40[\"split\"].loc[df_40[\"member\"].isin(test_members)] = \"test\"\n",
    "\n",
    "len(training_UIDs), len(test_UIDs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(71791, 290206)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_UID_MID_train = df_UID_MID.loc[df_UID_MID[\"Uniprot ID\"].isin(training_UIDs)]\n",
    "df_UID_MID_test = df_UID_MID.loc[df_UID_MID[\"Uniprot ID\"].isin(test_UIDs)]\n",
    "len(df_UID_MID_test), len(df_UID_MID_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculating for every sequence in the validation and test set the maximum accuracy compared to sequences in the training set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2000\n",
      "4000\n",
      "5000\n",
      "8000\n",
      "14000\n",
      "17000\n",
      "18000\n",
      "23000\n",
      "24000\n",
      "26000\n",
      "27000\n",
      "33000\n",
      "37000\n",
      "41000\n",
      "45000\n",
      "51000\n",
      "54000\n",
      "59000\n",
      "67000\n",
      "68000\n",
      "70000\n",
      "87000\n",
      "90000\n",
      "97000\n",
      "115000\n",
      "123000\n",
      "129000\n",
      "132000\n",
      "149000\n",
      "153000\n",
      "158000\n",
      "159000\n",
      "167000\n",
      "168000\n",
      "177000\n",
      "178000\n",
      "192000\n",
      "198000\n",
      "202000\n",
      "209000\n",
      "216000\n",
      "217000\n",
      "218000\n",
      "219000\n",
      "220000\n",
      "223000\n",
      "224000\n",
      "229000\n"
     ]
    }
   ],
   "source": [
    "df_80[\"identity\"] = np.nan\n",
    "df_80[\"identity\"].loc[df_80[\"split\"].isin([\"test\"])] =  \"60-80%\"\n",
    "\n",
    "test_indices = list(df_80.loc[~pd.isnull(df_80[\"identity\"])].index)\n",
    "\n",
    "\n",
    "for ind in test_indices:\n",
    "\n",
    "    member = df_80[\"member\"][ind]\n",
    "    cluster = list(df_40[\"cluster\"].loc[df_40[\"member\"] == member])[0]\n",
    "    cluster_splits = list(df_40[\"split\"].loc[df_40[\"cluster\"] == cluster].loc[df_40[\"evidence\"] == \"exp\"])\n",
    "    if not \"train\" in cluster_splits:\n",
    "        df_80[\"identity\"][ind] = \"<40%\"\n",
    "    else:\n",
    "        cluster = list(df_60[\"cluster\"].loc[df_60[\"member\"] == member])[0]\n",
    "        cluster_splits = list(df_60[\"split\"].loc[df_60[\"cluster\"] == cluster].loc[df_60[\"evidence\"] == \"exp\"])\n",
    "        if not \"train\" in cluster_splits:\n",
    "            df_80[\"identity\"][ind] = \"40-60%\"\n",
    "            \n",
    "    if ind % 1000 == 0:\n",
    "        print(ind)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_80.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\",\n",
    "                          \"df_80.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>member</th>\n",
       "      <th>split</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>59618</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>17477</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>193350</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>202039</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>31895</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230797</th>\n",
       "      <td>133168</td>\n",
       "      <td>102904</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230798</th>\n",
       "      <td>133169</td>\n",
       "      <td>168585</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230799</th>\n",
       "      <td>133170</td>\n",
       "      <td>39204</td>\n",
       "      <td>test</td>\n",
       "      <td>&lt;40%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230800</th>\n",
       "      <td>133171</td>\n",
       "      <td>42158</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230801</th>\n",
       "      <td>133172</td>\n",
       "      <td>99620</td>\n",
       "      <td>train</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230802 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cluster  member  split identity\n",
       "0             0   59618  train      NaN\n",
       "1             1   17477  train      NaN\n",
       "2             2  193350  train      NaN\n",
       "3             3  202039  train      NaN\n",
       "4             3   31895  train      NaN\n",
       "...         ...     ...    ...      ...\n",
       "230797   133168  102904  train      NaN\n",
       "230798   133169  168585  train      NaN\n",
       "230799   133170   39204   test     <40%\n",
       "230800   133171   42158  train      NaN\n",
       "230801   133172   99620  train      NaN\n",
       "\n",
       "[230802 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_80"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.DataFrame({\"member\": df_80[\"member\"], \"identity\": df_80[\"identity\"]})\n",
    "UNIPROT_df[\"member\"] = [str(ind) for ind in UNIPROT_df.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>cluster</th>\n",
       "      <th>member</th>\n",
       "      <th>identity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G3QP07</td>\n",
       "      <td>MAAGAGTAGLASGPGVVRDPAASQPRKRPGREGGEGARRSDTMAGG...</td>\n",
       "      <td>43869.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9KNU3</td>\n",
       "      <td>MRNTMFTSKEGQTIPQVTFPTRQGDAWVNVTSDELFKGKTVIVFSL...</td>\n",
       "      <td>109863.0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H3HBD2</td>\n",
       "      <td>MYPSVALLLWLWGAGVALQVIGLLGCLVEQNDTSDCASTARDRRTS...</td>\n",
       "      <td>17764.0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A3B6C0N2</td>\n",
       "      <td>MEYHRVVSLVAVVVVLLRRWPALSSAQAPVSRTITVDSHGGGDFSS...</td>\n",
       "      <td>88170.0</td>\n",
       "      <td>3</td>\n",
       "      <td>&lt;40%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A1Z5RRB7</td>\n",
       "      <td>MDMPSHTHSQLCESKALVASYTQEARKRNQQHNMASKPGPLSRWPW...</td>\n",
       "      <td>28018.0</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230809</th>\n",
       "      <td>A0A072VJI1</td>\n",
       "      <td>MELSAVTLGGKGSSLSSSAVYATAIGKSQIKIDSSALDRLTSPPSS...</td>\n",
       "      <td>15606.0</td>\n",
       "      <td>230809</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230810</th>\n",
       "      <td>D7SVZ9</td>\n",
       "      <td>MFIESFKVESPNVKYTENEIHSVYDYETTELVHENRNGTYQWVVKP...</td>\n",
       "      <td>34817.0</td>\n",
       "      <td>230810</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230811</th>\n",
       "      <td>A0A2C9WL07</td>\n",
       "      <td>MAPISILLFSSILLFSASSTGRALSFNYYEKTCPDVELIVTNAVKN...</td>\n",
       "      <td>93693.0</td>\n",
       "      <td>230811</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230812</th>\n",
       "      <td>F4ICB6</td>\n",
       "      <td>MGCVNSKQTVSVTPAIDHSGVFRDNVCSGSGRIVVEDLPPVTETKL...</td>\n",
       "      <td>22447.0</td>\n",
       "      <td>230812</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230813</th>\n",
       "      <td>B5LAU7</td>\n",
       "      <td>MAAAATCAFFPTGNPPSDSGAKSSGNLGGGSVPGSIDARGLNNVKK...</td>\n",
       "      <td>66315.0</td>\n",
       "      <td>230813</td>\n",
       "      <td>60-80%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230814 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Sequence  \\\n",
       "0           G3QP07  MAAGAGTAGLASGPGVVRDPAASQPRKRPGREGGEGARRSDTMAGG...   \n",
       "1           Q9KNU3  MRNTMFTSKEGQTIPQVTFPTRQGDAWVNVTSDELFKGKTVIVFSL...   \n",
       "2           H3HBD2  MYPSVALLLWLWGAGVALQVIGLLGCLVEQNDTSDCASTARDRRTS...   \n",
       "3       A0A3B6C0N2  MEYHRVVSLVAVVVVLLRRWPALSSAQAPVSRTITVDSHGGGDFSS...   \n",
       "4       A0A1Z5RRB7  MDMPSHTHSQLCESKALVASYTQEARKRNQQHNMASKPGPLSRWPW...   \n",
       "...            ...                                                ...   \n",
       "230809  A0A072VJI1  MELSAVTLGGKGSSLSSSAVYATAIGKSQIKIDSSALDRLTSPPSS...   \n",
       "230810      D7SVZ9  MFIESFKVESPNVKYTENEIHSVYDYETTELVHENRNGTYQWVVKP...   \n",
       "230811  A0A2C9WL07  MAPISILLFSSILLFSASSTGRALSFNYYEKTCPDVELIVTNAVKN...   \n",
       "230812      F4ICB6  MGCVNSKQTVSVTPAIDHSGVFRDNVCSGSGRIVVEDLPPVTETKL...   \n",
       "230813      B5LAU7  MAAAATCAFFPTGNPPSDSGAKSSGNLGGGSVPGSIDARGLNNVKK...   \n",
       "\n",
       "         cluster  member identity  \n",
       "0        43869.0       0      NaN  \n",
       "1       109863.0       1      NaN  \n",
       "2        17764.0       2      NaN  \n",
       "3        88170.0       3     <40%  \n",
       "4        28018.0       4      NaN  \n",
       "...          ...     ...      ...  \n",
       "230809   15606.0  230809      NaN  \n",
       "230810   34817.0  230810      NaN  \n",
       "230811   93693.0  230811      NaN  \n",
       "230812   22447.0  230812      NaN  \n",
       "230813   66315.0  230813   60-80%  \n",
       "\n",
       "[230814 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df = UNIPROT_df.merge(df2, on = \"member\", how = \"left\")\n",
    "        \n",
    "UNIPROT_df.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\",\n",
    "                          \"Uniprot_df_with_seq_identities.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8406, 12498, 24598)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(UNIPROT_df.loc[UNIPROT_df[\"identity\"] == \"60-80%\"]), len(UNIPROT_df.loc[UNIPROT_df[\"identity\"] == \"40-60%\"]), len(UNIPROT_df.loc[UNIPROT_df[\"identity\"] == \"<40%\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)(iii) For every enzyme in the test set with a sequence-identity of 60-80% we want to find out: How many enzymes with a sequence identity over 40% exist in the training set?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1294"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UNIPROT_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\",\n",
    "                          \"Uniprot_df_with_seq_identities.pkl\"))\n",
    "df_UID_MID = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID.pkl\"))\n",
    "df_UID_MID_train = df_UID_MID.loc[df_UID_MID[\"Uniprot ID\"].isin(training_UIDs)]\n",
    "df_UID_MID_test = df_UID_MID.loc[df_UID_MID[\"Uniprot ID\"].isin(test_UIDs)]\n",
    "df_UID_MID_train_exp = df_UID_MID_train.loc[df_UID_MID_train[\"evidence\"] == \"exp\"]\n",
    "exp_UIDs = list(set(df_UID_MID[\"Uniprot ID\"].loc[df_UID_MID[\"evidence\"] == \"exp\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df[\"n_similiar_train_enz\"] = np.nan\n",
    "UIDs_indices = list(UNIPROT_df[\"Uniprot ID\"].loc[UNIPROT_df[\"identity\"] == \"60-80%\"].index)\n",
    "\n",
    "for ind in UIDs_indices:\n",
    "    UID = UNIPROT_df[\"Uniprot ID\"][ind]\n",
    "    if UID in exp_UIDs:\n",
    "        cluster = list(df_40[\"cluster\"].loc[df_40[\"member\"] == str(ind)])[0]\n",
    "        cluster_members = list(df_40[\"member\"].loc[df_40[\"cluster\"] == cluster])\n",
    "        cluster_UIDs = [UNIPROT_df[\"Uniprot ID\"][int(member)] for member in cluster_members]\n",
    "        UNIPROT_df[\"n_similiar_train_enz\"][ind] = len(df_UID_MID_train_exp.loc[df_UID_MID_train_exp[\"Uniprot ID\"].isin(cluster_UIDs)])\n",
    "        \n",
    "UNIPROT_df.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\",\n",
    "                          \"Uniprot_df_with_seq_identities_V2.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Calculating enzyme and substrate representations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"Uniprot_df_with_seq_identities.pkl\"))\n",
    "df_UID_MID = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Calculate extended-connectivity fingerprints (ECFPs) for substrates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>substrate ID</th>\n",
       "      <th>ECFP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEBI:18378</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEBI:58282</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEBI:17071</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEBI:57955</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C21650</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1757</th>\n",
       "      <td>CHEBI:58628</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1758</th>\n",
       "      <td>CHEBI:16737</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1759</th>\n",
       "      <td>CHEBI:30909</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1760</th>\n",
       "      <td>CHEBI:61519</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1761</th>\n",
       "      <td>CHEBI:58189</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1762 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     substrate ID ECFP\n",
       "0     CHEBI:18378     \n",
       "1     CHEBI:58282     \n",
       "2     CHEBI:17071     \n",
       "3     CHEBI:57955     \n",
       "4          C21650     \n",
       "...           ...  ...\n",
       "1757  CHEBI:58628     \n",
       "1758  CHEBI:16737     \n",
       "1759  CHEBI:30909     \n",
       "1760  CHEBI:61519     \n",
       "1761  CHEBI:58189     \n",
       "\n",
       "[1762 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mol_folder = \"C:\\\\Users\\\\alexk\\\\mol-files\\\\\"\n",
    "mol_IDs = list(set(df_UID_MID[\"molecule ID\"]))\n",
    "\n",
    "df_ecfps = pd.DataFrame(data = {\"substrate ID\" : mol_IDs})\n",
    "df_ecfps[\"ECFP\"] = \"\"\n",
    "df_ecfps"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating txt file with all ChEBI IDs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ChEBI_IDs = list(set(df_UID_MID[\"molecule ID\"]))\n",
    "f = open(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"ChEBI_IDs.txt\"),\"w\")\n",
    "for ID in list(set(ChEBI_IDs)):\n",
    "    if ID[:2] == \"CH\":\n",
    "        f.write(str(ID.replace(\"CHEBI\", \"ChEBI\")) + \"\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we map ChEBI IDs to InChI strings using https://metacyc.org/metabolite-translation-service.shtml:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>Result</th>\n",
       "      <th>BioCyc Common-Name</th>\n",
       "      <th>BioCyc</th>\n",
       "      <th>Kegg</th>\n",
       "      <th>ChEBI</th>\n",
       "      <th>PubChem</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>ChemSpider</th>\n",
       "      <th>MetaboLights</th>\n",
       "      <th>MetaNetX</th>\n",
       "      <th>BiGG</th>\n",
       "      <th>Seed</th>\n",
       "      <th>Inchi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ChEBI:132511</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ChEBI:11805</td>\n",
       "      <td>success</td>\n",
       "      <td>3-hydroxy-isobutanoate</td>\n",
       "      <td>3-HYDROXY-ISOBUTYRATE</td>\n",
       "      <td>C01188</td>\n",
       "      <td>11805.0</td>\n",
       "      <td>11966314</td>\n",
       "      <td>HMDB62640</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MNXM396</td>\n",
       "      <td></td>\n",
       "      <td>cpd00876</td>\n",
       "      <td>InChI=1S/C4H8O3/c1-3(2-5)4(6)7/h3,5H,2H2,1H3,(...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ChEBI:17568</td>\n",
       "      <td>success</td>\n",
       "      <td>uracil</td>\n",
       "      <td>URACIL</td>\n",
       "      <td>C00106</td>\n",
       "      <td>17568.0</td>\n",
       "      <td>1174</td>\n",
       "      <td>HMDB00300</td>\n",
       "      <td>1141</td>\n",
       "      <td>MTBLC17568</td>\n",
       "      <td>MNXM158</td>\n",
       "      <td>ura</td>\n",
       "      <td>cpd00092</td>\n",
       "      <td>InChI=1S/C4H4N2O2/c7-3-1-2-5-4(8)6-3/h1-2H,(H2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ChEBI:17237</td>\n",
       "      <td>success</td>\n",
       "      <td>crotono-betaine</td>\n",
       "      <td>CROTONO-BETAINE</td>\n",
       "      <td>C04114</td>\n",
       "      <td>17237.0</td>\n",
       "      <td>5462194</td>\n",
       "      <td></td>\n",
       "      <td>4575319</td>\n",
       "      <td></td>\n",
       "      <td>MNXM1607</td>\n",
       "      <td>ctbt</td>\n",
       "      <td>cpd02543</td>\n",
       "      <td>InChI=1S/C7H13NO2/c1-8(2,3)6-4-5-7(9)10/h4-5H,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ChEBI:35121</td>\n",
       "      <td>success</td>\n",
       "      <td>(2R,3S)-3-isopropylmalate</td>\n",
       "      <td>2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE</td>\n",
       "      <td>C04411</td>\n",
       "      <td>35121.0</td>\n",
       "      <td>6857402</td>\n",
       "      <td>HMDB12156</td>\n",
       "      <td>5256741</td>\n",
       "      <td></td>\n",
       "      <td>MNXM891</td>\n",
       "      <td>3c2hmp</td>\n",
       "      <td>cpd02693</td>\n",
       "      <td>InChI=1S/C7H12O5/c1-3(2)4(6(9)10)5(8)7(11)12/h...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Input   Result         BioCyc Common-Name  \\\n",
       "0  ChEBI:132511  unknown                        NaN   \n",
       "1   ChEBI:11805  success     3-hydroxy-isobutanoate   \n",
       "2   ChEBI:17568  success                     uracil   \n",
       "3   ChEBI:17237  success            crotono-betaine   \n",
       "4   ChEBI:35121  success  (2R,3S)-3-isopropylmalate   \n",
       "\n",
       "                                    BioCyc    Kegg    ChEBI   PubChem  \\\n",
       "0                                      NaN     NaN      NaN       NaN   \n",
       "1                    3-HYDROXY-ISOBUTYRATE  C01188  11805.0  11966314   \n",
       "2                                   URACIL  C00106  17568.0      1174   \n",
       "3                          CROTONO-BETAINE  C04114  17237.0   5462194   \n",
       "4  2-D-THREO-HYDROXY-3-CARBOXY-ISOCAPROATE  C04411  35121.0   6857402   \n",
       "\n",
       "        HMDB ChemSpider MetaboLights  MetaNetX    BiGG      Seed  \\\n",
       "0        NaN        NaN          NaN       NaN     NaN       NaN   \n",
       "1  HMDB62640                           MNXM396          cpd00876   \n",
       "2  HMDB00300       1141   MTBLC17568   MNXM158     ura  cpd00092   \n",
       "3               4575319               MNXM1607    ctbt  cpd02543   \n",
       "4  HMDB12156    5256741                MNXM891  3c2hmp  cpd02693   \n",
       "\n",
       "                                               Inchi  \n",
       "0                                                NaN  \n",
       "1  InChI=1S/C4H8O3/c1-3(2-5)4(6)7/h3,5H,2H2,1H3,(...  \n",
       "2  InChI=1S/C4H4N2O2/c7-3-1-2-5-4(8)6-3/h1-2H,(H2...  \n",
       "3  InChI=1S/C7H13NO2/c1-8(2,3)6-4-5-7(9)10/h4-5H,...  \n",
       "4  InChI=1S/C7H12O5/c1-3(2)4(6(9)10)5(8)7(11)12/h...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_chebi_to_inchi = pd.read_csv(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"chebiID_to_inchi.tsv\"), sep = \"\\t\")\n",
    "df_chebi_to_inchi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mapping InChI string to DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ind in df_ecfps.index:\n",
    "    met_ID = df_ecfps[\"substrate ID\"][ind]\n",
    "    is_CHEBI_ID = (met_ID[0:5] == \"CHEBI\")\n",
    "    \n",
    "    \n",
    "    if is_CHEBI_ID:\n",
    "        try:\n",
    "            ID = int(met_ID.split(\" \")[0].split(\":\")[-1])\n",
    "            Inchi = list(df_chebi_to_inchi[\"Inchi\"].loc[df_chebi_to_inchi[\"ChEBI\"] == float(ID)])[0]\n",
    "            if not pd.isnull(Inchi):\n",
    "                mol = Chem.inchi.MolFromInchi(Inchi)\n",
    "        except IndexError:\n",
    "            mol = None\n",
    "        \n",
    "    else:\n",
    "        try:\n",
    "            mol = Chem.MolFromMolFile(mol_folder +  \"/mol-files/\" + met_ID + '.mol')\n",
    "        except OSError:\n",
    "            None\n",
    "            \n",
    "    if mol is not None:\n",
    "        ecfp = AllChem.GetMorganFingerprintAsBitVect(mol, 3, nBits=1024).ToBitString()\n",
    "        df_ecfps[\"ECFP\"][ind] = ecfp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ecfps = df_ecfps.loc[df_ecfps[\"ECFP\"] !=\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ecfps.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"df_ecfps.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (b) Calculating enzyme representations with the ESM-1b model:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ESM-1b model takes a FASTA file with the enzymes' amino acid sequences as its input. Creating a FASTA file with enzyme sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "ofile = open(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"all_sequences.fasta\"), \"w\")\n",
    "for ind in UNIPROT_df.index:\n",
    "    seq = UNIPROT_df[\"Sequence\"][ind]\n",
    "    if not pd.isnull(seq):\n",
    "        seq_end = seq.find(\"#\")\n",
    "        seq = seq[:seq_end]\n",
    "        ofile.write(\">\" + str(ind) + \"\\n\" + seq[:1018]  + \"\\n\")\n",
    "ofile.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To calculate the ESM-1b vectors, we used the model and code provided by the Facebook Research team: https://github.com/facebookresearch/esm. The following command line was used to calculate the representations:\n",
    "\n",
    "<span style=\"color:red\"> \n",
    "python extract.py esm1b_t33_650M_UR50S \\path_to_fasta_file\\all_sequences.fasta \\path_to_store_representations\\all_sequences_esm1b_reps--repr_layers 33 --include mean </span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading ESM-1b-vectors and storing them in the UniProt DataFrame. All representations were merged into one dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>cluster</th>\n",
       "      <th>identity</th>\n",
       "      <th>ESM1b</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G3QP07</td>\n",
       "      <td>MAAGAGTAGLASGPGVVRDPAASQPRKRPGREGGEGARRSDTMAGG...</td>\n",
       "      <td>43869.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[-0.024416907, 0.18361057, 0.05229881, 0.04472...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q9KNU3</td>\n",
       "      <td>MRNTMFTSKEGQTIPQVTFPTRQGDAWVNVTSDELFKGKTVIVFSL...</td>\n",
       "      <td>109863.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[-0.10067828, 0.1986571, -0.21036363, 0.013421...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H3HBD2</td>\n",
       "      <td>MYPSVALLLWLWGAGVALQVIGLLGCLVEQNDTSDCASTARDRRTS...</td>\n",
       "      <td>17764.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0.040806588, 0.1870469, 0.11258543, 0.0708398...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A3B6C0N2</td>\n",
       "      <td>MEYHRVVSLVAVVVVLLRRWPALSSAQAPVSRTITVDSHGGGDFSS...</td>\n",
       "      <td>88170.0</td>\n",
       "      <td>60-80%</td>\n",
       "      <td>[0.07788918, 0.11309477, 0.021622797, 0.094022...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A1Z5RRB7</td>\n",
       "      <td>MDMPSHTHSQLCESKALVASYTQEARKRNQQHNMASKPGPLSRWPW...</td>\n",
       "      <td>28018.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0.04624611, 0.27114883, 0.20789135, -0.034673...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230809</th>\n",
       "      <td>A0A072VJI1</td>\n",
       "      <td>MELSAVTLGGKGSSLSSSAVYATAIGKSQIKIDSSALDRLTSPPSS...</td>\n",
       "      <td>15606.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[-0.04588662, 0.27346706, 0.12282053, 0.027531...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230810</th>\n",
       "      <td>D7SVZ9</td>\n",
       "      <td>MFIESFKVESPNVKYTENEIHSVYDYETTELVHENRNGTYQWVVKP...</td>\n",
       "      <td>34817.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0.0014713518, 0.20142357, 0.077046216, -0.077...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230811</th>\n",
       "      <td>A0A2C9WL07</td>\n",
       "      <td>MAPISILLFSSILLFSASSTGRALSFNYYEKTCPDVELIVTNAVKN...</td>\n",
       "      <td>93693.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[-0.041781973, 0.19652756, 0.077766806, 0.1401...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230812</th>\n",
       "      <td>F4ICB6</td>\n",
       "      <td>MGCVNSKQTVSVTPAIDHSGVFRDNVCSGSGRIVVEDLPPVTETKL...</td>\n",
       "      <td>22447.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[0.052631423, 0.29968897, -0.037843555, -0.007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230813</th>\n",
       "      <td>B5LAU7</td>\n",
       "      <td>MAAAATCAFFPTGNPPSDSGAKSSGNLGGGSVPGSIDARGLNNVKK...</td>\n",
       "      <td>66315.0</td>\n",
       "      <td>60-80%</td>\n",
       "      <td>[-0.037645016, 0.23983482, -0.044912808, 0.028...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>230814 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Uniprot ID                                           Sequence  \\\n",
       "0           G3QP07  MAAGAGTAGLASGPGVVRDPAASQPRKRPGREGGEGARRSDTMAGG...   \n",
       "1           Q9KNU3  MRNTMFTSKEGQTIPQVTFPTRQGDAWVNVTSDELFKGKTVIVFSL...   \n",
       "2           H3HBD2  MYPSVALLLWLWGAGVALQVIGLLGCLVEQNDTSDCASTARDRRTS...   \n",
       "3       A0A3B6C0N2  MEYHRVVSLVAVVVVLLRRWPALSSAQAPVSRTITVDSHGGGDFSS...   \n",
       "4       A0A1Z5RRB7  MDMPSHTHSQLCESKALVASYTQEARKRNQQHNMASKPGPLSRWPW...   \n",
       "...            ...                                                ...   \n",
       "230809  A0A072VJI1  MELSAVTLGGKGSSLSSSAVYATAIGKSQIKIDSSALDRLTSPPSS...   \n",
       "230810      D7SVZ9  MFIESFKVESPNVKYTENEIHSVYDYETTELVHENRNGTYQWVVKP...   \n",
       "230811  A0A2C9WL07  MAPISILLFSSILLFSASSTGRALSFNYYEKTCPDVELIVTNAVKN...   \n",
       "230812      F4ICB6  MGCVNSKQTVSVTPAIDHSGVFRDNVCSGSGRIVVEDLPPVTETKL...   \n",
       "230813      B5LAU7  MAAAATCAFFPTGNPPSDSGAKSSGNLGGGSVPGSIDARGLNNVKK...   \n",
       "\n",
       "         cluster identity                                              ESM1b  \n",
       "0        43869.0      NaN  [-0.024416907, 0.18361057, 0.05229881, 0.04472...  \n",
       "1       109863.0      NaN  [-0.10067828, 0.1986571, -0.21036363, 0.013421...  \n",
       "2        17764.0      NaN  [0.040806588, 0.1870469, 0.11258543, 0.0708398...  \n",
       "3        88170.0   60-80%  [0.07788918, 0.11309477, 0.021622797, 0.094022...  \n",
       "4        28018.0      NaN  [0.04624611, 0.27114883, 0.20789135, -0.034673...  \n",
       "...          ...      ...                                                ...  \n",
       "230809   15606.0      NaN  [-0.04588662, 0.27346706, 0.12282053, 0.027531...  \n",
       "230810   34817.0      NaN  [0.0014713518, 0.20142357, 0.077046216, -0.077...  \n",
       "230811   93693.0      NaN  [-0.041781973, 0.19652756, 0.077766806, 0.1401...  \n",
       "230812   22447.0      NaN  [0.052631423, 0.29968897, -0.037843555, -0.007...  \n",
       "230813   66315.0   60-80%  [-0.037645016, 0.23983482, -0.044912808, 0.028...  \n",
       "\n",
       "[230814 rows x 5 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UNIPROT_df[\"ESM1b\"] = \"\"\n",
    "\n",
    "rep_dict = torch.load(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"all_sequences_enz_sub.pt\"))\n",
    "\n",
    "for ind in UNIPROT_df.index:\n",
    "    try:\n",
    "        UNIPROT_df[\"ESM1b\"][ind] = rep_dict[str(ind) +\".pt\"]\n",
    "    except:\n",
    "        print(ind)\n",
    "UNIPROT_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"Uniprot_df_with_ESM1b.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Adding negative data points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"Uniprot_df_with_ESM1b.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only using UniProt IDs with ESM1b vectors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "UIDs_with_rep = list(set(UNIPROT_df[\"Uniprot ID\"].loc[UNIPROT_df[\"ESM1b\"] != \"\"]))\n",
    "\n",
    "df_UID_MID_train = df_UID_MID_train.loc[df_UID_MID_train[\"Uniprot ID\"].isin(UIDs_with_rep)]\n",
    "df_UID_MID_test = df_UID_MID_test.loc[df_UID_MID_test[\"Uniprot ID\"].isin(UIDs_with_rep)]\n",
    "\n",
    "df_UID_MID_train = df_UID_MID_train.sample(frac = 1, random_state = 1).reset_index(drop = True)\n",
    "df_UID_MID_test = df_UID_MID_test.sample(frac = 1, random_state = 1).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mapping ECFPs to DataFrames:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ecfps = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"df_ecfps.pkl\"))\n",
    "\n",
    "df_UID_MID_train[\"ECFP\"] = \"\"\n",
    "df_UID_MID_test[\"ECFP\"] = \"\"\n",
    "for ind in df_ecfps.index:\n",
    "    sub_ID, ecfp = df_ecfps[\"substrate ID\"][ind], df_ecfps[\"ECFP\"][ind]\n",
    "    df_UID_MID_train[\"ECFP\"].loc[df_UID_MID_train[\"molecule ID\"] == sub_ID] = ecfp\n",
    "    df_UID_MID_test[\"ECFP\"].loc[df_UID_MID_test[\"molecule ID\"] == sub_ID] = ecfp\n",
    "    \n",
    "df_UID_MID_test = df_UID_MID_test.loc[df_UID_MID_test[\"ECFP\"] != \"\"]\n",
    "df_UID_MID_train = df_UID_MID_train.loc[df_UID_MID_train[\"ECFP\"] != \"\"]\n",
    "\n",
    "df_UID_MID_train.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\", \"df_UID_MID_train.pkl\"))\n",
    "df_UID_MID_test.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID_test.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Only using enzymes with more than two substrate:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\", \"df_UID_MID_train.pkl\"))\n",
    "df_UID_MID_test = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID_test.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_UIDs = list(df_UID_MID_train[\"Uniprot ID\"])\n",
    "UIDs = list(set(all_UIDs)) \n",
    "n_pos = [UIDs.count(UID) for UID in UIDs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "promiscous_enzymes = np.array(UIDs)[list(np.array(n_pos) >= 3)]\n",
    "df_UID_MID_train = df_UID_MID_train.loc[df_UID_MID_train[\"Uniprot ID\"].isin(promiscous_enzymes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_UIDs = list(df_UID_MID_test[\"Uniprot ID\"])\n",
    "UIDs = list(set(all_UIDs)) \n",
    "n_pos = [UIDs.count(UID) for UID in UIDs]\n",
    "promiscous_enzymes = np.array(UIDs)[list(np.array(n_pos) >= 3)]\n",
    "df_UID_MID_test = df_UID_MID_test.loc[df_UID_MID_test[\"Uniprot ID\"].isin(promiscous_enzymes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.concat([df_UID_MID_train, df_UID_MID_test], ignore_index = True)\n",
    "\n",
    "df_exp = df_all.loc[df_all[\"evidence\"] == \"exp\"]\n",
    "df_phylo = df_all.loc[df_all[\"evidence\"] == \"phylo\"]\n",
    "\n",
    "print(\"We have %s entries with phylogenetic evidence and %s entries with experimental evidence\" % (len(df_phylo), len(df_exp)))\n",
    "\n",
    "print(\"\\n experimental dataset:\")\n",
    "print(\"Number of different enzymes: %s, Number of different substrates: %s\"\n",
    "      % (len(set(df_exp[\"Uniprot ID\"])), len(set(df_exp[\"molecule ID\"]))) )\n",
    "\n",
    "print(\"\\n phylogenetic dataset:\")\n",
    "print(\"Number of different enzymes: %s, Number of different substrates: %s\"\n",
    "      % (len(set(df_phylo[\"Uniprot ID\"])), len(set(df_phylo[\"molecule ID\"]))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Creating negative data points for the training and test set:\n",
    "To assign the negative data points, we will choose similar metabolites compared to the substrate of the psoitive datapoints. Therefore, we are first creating a similarity matrix for all metabolites in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_chebi_to_inchi = pd.read_csv(join(CURRENT_DIR, \"..\" ,\"data\", \"substrate_data\", \"chebiID_to_inchi.tsv\"), sep = \"\\t\")\n",
    "mol_folder = \"C:\\\\Users\\\\alexk\\\\mol-files\\\\\"\n",
    "\n",
    "def get_mol(met_ID):\n",
    "    is_CHEBI_ID = (met_ID[0:5] == \"CHEBI\")\n",
    "    is_InChI = (met_ID[0:5] == \"InChI\")\n",
    "    if is_CHEBI_ID:\n",
    "        try:\n",
    "            ID = int(met_ID.split(\" \")[0].split(\":\")[-1])\n",
    "            Inchi = list(df_chebi_to_inchi[\"Inchi\"].loc[df_chebi_to_inchi[\"ChEBI\"] == float(ID)])[0]\n",
    "            mol = Chem.inchi.MolFromInchi(Inchi)\n",
    "        except:\n",
    "            mol = None     \n",
    "    elif is_InChI:\n",
    "        try:\n",
    "            mol = Chem.inchi.MolFromInchi(met_ID)\n",
    "        except:\n",
    "            mol = None\n",
    "        \n",
    "    else:\n",
    "        try:\n",
    "            mol = Chem.MolFromMolFile(mol_folder +  \"mol-files\\\\\" + met_ID + '.mol')\n",
    "        except OSError:\n",
    "            mol = None\n",
    "            \n",
    "    return(mol)\n",
    "\n",
    "def drop_samples_without_mol_file(df):\n",
    "    droplist = []\n",
    "    for ind in df.index:\n",
    "        if get_mol(met_ID = df[\"molecule ID\"][ind]) is None:\n",
    "            droplist.append(ind)\n",
    "\n",
    "    df.drop(droplist, inplace = True)\n",
    "    return(df)\n",
    "\n",
    "def get_metabolites_and_similarities(df):\n",
    "    df_metabolites = pd.DataFrame(data = {\"ECFP\": df[\"ECFP\"], \"ID\": df[\"molecule ID\"]})\n",
    "    df_metabolites = df_metabolites.drop_duplicates()\n",
    "    df_metabolites.reset_index(inplace = True, drop = True)\n",
    "\n",
    "\n",
    "    ms = [get_mol(met_ID = df_metabolites[\"ID\"][ind]) for ind in df_metabolites.index]\n",
    "    fps = [Chem.RDKFingerprint(x) for x in ms]\n",
    "\n",
    "    similarity_matrix = np.zeros((len(ms), len(ms)))\n",
    "    for i in range(len(ms)):\n",
    "        for j in range(len(ms)):\n",
    "            similarity_matrix[i,j] = DataStructs.FingerprintSimilarity(fps[i],fps[j])\n",
    "            \n",
    "    return(df_metabolites, similarity_matrix)\n",
    "\n",
    "\n",
    "\n",
    "def get_valid_list(met_ID, UID, forbidden_metabolites, df_metabolites, similarity_matrix, lower_bound =0.7, upper_bound =0.9):\n",
    "    binding_met_IDs = list(df_UID_MID[\"molecule ID\"].loc[df_UID_MID[\"Uniprot ID\"] == UID])\n",
    "    k = df_metabolites.loc[df_metabolites[\"ID\"] == met_ID].index[0]\n",
    "\n",
    "    similarities = similarity_matrix[k,:]\n",
    "    selection = (similarities< upper_bound) * (similarities >lower_bound) \n",
    "    metabolites = list(df_metabolites[\"ID\"].loc[selection])\n",
    "    \n",
    "    no_mets = list(set(binding_met_IDs + forbidden_metabolites))\n",
    "    \n",
    "    metabolites = [met for met in metabolites if (met not in no_mets)]\n",
    "    return(metabolites)\n",
    "\n",
    "\n",
    "def create_negative_samples(df, df_metabolites, similarity_matrix):\n",
    "    start = time.time()\n",
    "    UID_list = []\n",
    "    MID_list = []\n",
    "    Type_list = []\n",
    "    forbidden_mets = []\n",
    "\n",
    "    for ind in df.index:\n",
    "        if ind % 100 ==0:\n",
    "            print(ind)\n",
    "            print(\"Time: %s [min]\" % np.round(float((time.time()-start)/60),2))\n",
    "\n",
    "            df2 = pd.DataFrame(data = {\"Uniprot ID\": UID_list, \"molecule ID\" : MID_list, \"type\" : Type_list})\n",
    "            df2[\"Binding\"] = 0\n",
    "            df = pd.concat([df, df2], ignore_index=True)\n",
    "\n",
    "            UID_list, MID_list, Type_list = [], [], []\n",
    "\n",
    "            forbidden_mets_old = forbidden_mets.copy()\n",
    "            all_mets = list(set(df[\"molecule ID\"]))\n",
    "            all_mets = [met for met in all_mets if not met in forbidden_mets_old]\n",
    "            forbidden_mets = list(set([met for met in all_mets if \n",
    "                                       (np.mean(df[\"Binding\"].loc[df[\"molecule ID\"] == met]) < 1/4)]))\n",
    "            forbidden_mets = forbidden_mets + forbidden_mets_old\n",
    "            print(len(forbidden_mets))\n",
    "\n",
    "        UID = df[\"Uniprot ID\"][ind]\n",
    "        Type = df[\"type\"][ind]\n",
    "        met_ID = df[\"molecule ID\"][ind]\n",
    "\n",
    "        metabolites = get_valid_list(met_ID = met_ID, UID = UID, forbidden_metabolites= forbidden_mets,\n",
    "                                     df_metabolites = df_metabolites, similarity_matrix = similarity_matrix,\n",
    "                                     lower_bound =0.7, upper_bound =0.95)\n",
    "        lower_bound = 0.7\n",
    "        while len(metabolites) < 2:\n",
    "            lower_bound = lower_bound - 0.2\n",
    "            metabolites = get_valid_list(met_ID = met_ID, UID = UID, forbidden_metabolites= forbidden_mets,\n",
    "                                     df_metabolites = df_metabolites, similarity_matrix = similarity_matrix,\n",
    "                                     lower_bound =lower_bound, upper_bound =0.95)\n",
    "            if lower_bound <0:\n",
    "                break\n",
    "        \n",
    "        new_metabolites =  random.sample(metabolites, min(3,len(metabolites)))\n",
    "\n",
    "        for met in new_metabolites:\n",
    "            UID_list.append(UID), MID_list.append(met), Type_list.append(Type)\n",
    "\n",
    "    df2 = pd.DataFrame(data = {\"Uniprot ID\": UID_list, \"molecule ID\" : MID_list, \"type\" : Type_list})\n",
    "    df2[\"Binding\"] = 0\n",
    "\n",
    "    df = pd.concat([df, df2], ignore_index = True)\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)(i) Creating negative data points for the training set (experimental evidence):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_UID_MID_train_exp = df_UID_MID_train.loc[df_UID_MID_train[\"evidence\"] == \"exp\"]\n",
    "\n",
    "df_UID_MID_train_exp = drop_samples_without_mol_file(df = df_UID_MID_train_exp)\n",
    "#calculating similarity matrix for all metabolites in the set:\n",
    "df_metabolites_train, similarity_matrix_train = get_metabolites_and_similarities(df = df_UID_MID_train_exp)\n",
    "print(len(df_metabolites_train))\n",
    "\n",
    "df_UID_MID_train_exp[\"Binding\"] = 1\n",
    "df_UID_MID_train_exp.reset_index(inplace = True, drop = True)\n",
    "\n",
    "df_UID_MID_train_exp = create_negative_samples(df = df_UID_MID_train_exp, df_metabolites = df_metabolites_train,\n",
    "                                          similarity_matrix = similarity_matrix_train)\n",
    "df_UID_MID_train_exp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)(ii) Creating negative data points for the training set (phylogentical evidence):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train_phylo = df_UID_MID_train.loc[df_UID_MID_train[\"evidence\"] == \"phylo\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_UID_MID_train_phylo = drop_samples_without_mol_file(df = df_UID_MID_train_phylo)\n",
    "#calculating similarity matrix for all metabolites in the set:\n",
    "df_metabolites_train, similarity_matrix_train = get_metabolites_and_similarities(df = df_UID_MID_train_phylo)\n",
    "print(len(df_metabolites_train))\n",
    "\n",
    "df_UID_MID_train_phylo[\"Binding\"] = 1\n",
    "df_UID_MID_train_phylo.reset_index(inplace = True, drop = True)\n",
    "\n",
    "df_UID_MID_train_phylo = create_negative_samples(df = df_UID_MID_train_phylo, df_metabolites = df_metabolites_train,\n",
    "                                          similarity_matrix = similarity_matrix_train)\n",
    "df_UID_MID_train_phylo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (a)(iii) Creating negative data points for the test set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_test = df_UID_MID_test.loc[df_UID_MID_test[\"evidence\"] == \"exp\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_UID_MID_test = drop_samples_without_mol_file(df = df_UID_MID_test)\n",
    "#calculating similarity matrix for all metabolites in the set:\n",
    "df_metabolites_test, similarity_matrix_test = get_metabolites_and_similarities(df = df_UID_MID_test)\n",
    "print(len(df_metabolites_test))\n",
    "\n",
    "df_UID_MID_test[\"Binding\"] = 1\n",
    "df_UID_MID_test.reset_index(inplace = True, drop = True)\n",
    "\n",
    "df_UID_MID_test = create_negative_samples(df = df_UID_MID_test, df_metabolites = df_metabolites_test,\n",
    "                                          similarity_matrix = similarity_matrix_test)\n",
    "df_UID_MID_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train_phylo.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_train_phylo_promiscous.pkl\"))\n",
    "df_UID_MID_train_exp.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_train_exp_promiscous.pkl\"))\n",
    "df_UID_MID_test.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_test_exp_phylo_promiscous.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (b) Mapping ECFPs and ESM-1b-vectors to different splits:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train_phylo = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_train_phylo_promiscous.pkl\"))\n",
    "df_UID_MID_train_exp = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_train_exp_promiscous.pkl\"))\n",
    "df_UID_MID_test = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\",\"df_UID_MID_test_exp_phylo_promiscous.pkl\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train_exp[\"evidence\"] = \"exp\"\n",
    "df_UID_MID_train_phylo[\"evidence\"] = \"phylo\"\n",
    "df_UID_MID_train = pd.concat([df_UID_MID_train_exp, df_UID_MID_train_phylo], ignore_index = True)\n",
    "\n",
    "df_UID_MID_test[\"evidence\"] = \"exp\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(i) Mappings ECFPs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train.drop(columns = [\"ECFP\"], inplace = True)\n",
    "df_UID_MID_train = df_UID_MID_train.merge(df_ecfps, right_on = \"substrate ID\", left_on = \"molecule ID\", how = \"left\")\n",
    "\n",
    "df_UID_MID_test.drop(columns = [\"ECFP\"], inplace = True)\n",
    "df_UID_MID_test = df_UID_MID_test.merge(df_ecfps, right_on = \"substrate ID\", left_on = \"molecule ID\", how = \"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (b)(ii) Mappings ESM1b-vectors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = df_UID_MID_train\n",
    "df_test = df_UID_MID_test\n",
    "\n",
    "Uniprot_df = pd.DataFrame(data = {\"Uniprot ID\" : UNIPROT_df[\"Uniprot ID\"],\n",
    "                                 \"ESM1b\" : UNIPROT_df[\"ESM1b\"]})\n",
    "\n",
    "df_train = df_train.merge(Uniprot_df, on = \"Uniprot ID\", how = \"left\")\n",
    "df_test = df_test.merge(Uniprot_df, on = \"Uniprot ID\", how = \"left\")\n",
    "\n",
    "df_train.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"splits\", \"df_train_promiscous.pkl\"))\n",
    "df_test.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"splits\", \"df_test_promiscous.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Adding task-specific enzyme representations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (a) Creating input data for training the task-specific ESM1b vectors:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_UID_MID_train = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"enzyme_substrate_data\", \"df_UID_MID_train_promiscous.pkl\" ))\n",
    "df_UID_MID_test = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_substrate_data\", \"df_UID_MID_test_promiscous.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (b) Training the ESM1b model:\n",
    "Training is executed via terminal with the following command:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (c) Creating and mapping task-specific ESM1b vectors:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following command line was used to calculate the representations:\n",
    "\n",
    "<span style=\"color:red\"> \n",
    "python extract_ts.py esm1b_t33_650M_UR50S \\path_to_fasta_file\\all_sequences.fasta \\path_to_store_representations\\all_sequences_enz_sub_ts --repr_layers 33 --include mean </span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading ESM-1b-vectors and storing them in the UniProt DataFrame. All representations were merged into one dictionary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_df = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"Uniprot_df_with_ESM1b.pkl\"))\n",
    "UNIPROT_df.drop(columns = \"ESM1b\", inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep_dict = torch.load(join(CURRENT_DIR, \"..\" ,\"data\", \"enzyme_data\", \"all_sequences_enz_sub_ts.pt\"))\n",
    "\n",
    "UNIPROT_df[\"ESM1b_ts\"] = \"\"\n",
    "\n",
    "for ind in UNIPROT_df.index:\n",
    "    try:\n",
    "        UNIPROT_df[\"ESM1b_ts\"][ind] = rep_dict[str(ind) +\".pt\"]\n",
    "    except:\n",
    "        print(ind)\n",
    "UNIPROT_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mapping:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"splits\", \"df_train_promiscous.pkl\"))\n",
    "df_test = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"splits\", \"df_test_promiscous.pkl\"))\n",
    "\n",
    "Uniprot_df = pd.DataFrame(data = {\"Uniprot ID\" : UNIPROT_df[\"Uniprot ID\"],\n",
    "                                 \"ESM1b_ts\" : UNIPROT_df[\"ESM1b_ts\"]})\n",
    "\n",
    "df_train = df_train.merge(Uniprot_df, on = \"Uniprot ID\", how = \"left\")\n",
    "df_test = df_test.merge(Uniprot_df, on = \"Uniprot ID\", how = \"left\")\n",
    "\n",
    "df_train.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"splits\", \"df_train_with_ESM1b_ts_promiscous.pkl\"))\n",
    "df_test.to_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"splits\", \"df_test_with_ESM1b_ts_promiscous.pkl\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Adding task-specific metabolite representations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task-specific metabolite representations are created with the jupyter notebook \"Training Graph Neural Network\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\",\"splits\", \"df_train_with_ESM1b_ts_GNN.pkl\"))\n",
    "df_test = pd.read_pickle(join(CURRENT_DIR, \"..\" ,\"data\", \"splits\", \"df_test_with_ESM1b_ts_GNN.pkl\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
